---
title: Covariate Analysis of ROSMAP Reprocessed counts
author: 'JKG - Adapted from: Thanneer Perumal'
output: html_notebook
editor_options:
  chunk_output_type: console
---
Date of analysis update: "`r date()`"

| *Synapse ID* | *File Name* |
|  -----------------------------------------  |   ---------                      |
| syn3382527.7 | | 
| syn4300313.1 | | 
| ----Old Lengths syn8449369.2 | | 
| syn21156161 | | 
| syn7116000.1 | | 
| ----OLDCOUNTS syn8691134.1 | | 
| ----OLDCOUNTS syn21075911 | | 
| syn23593968 | | 
| ----SynID Translator | | 
| syn21075913 | | 
| syn3191087.3 | | 
| ----OLDMetrics syn8698240.2 Version 2 | | 
| syn21075912  | | 
| Full MetaData | |
| syn23573928 ||


### Load Libraries
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file
## Load required libraries
library(CovariateAnalysis) # get the package from devtools::install_github('th1vairam/CovariateAnalysis@dev')
library(data.table)
library(plyr)
library(tidyverse)
library(psych)
library(limma)
library(edgeR)
library(biomaRt)
library(RColorBrewer)
library(cqn)
library(knitr)
library(doParallel)
library(foreach)
library(githubr)
cl = makeCluster(detectCores()-2)
registerDoParallel(cl)
options(xtable.type="html")
knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

### Data download
Obtain count matrix and metadata from synapse
```{r download.data}
# Download expression data
COUNT_ID <- 'syn23593968';
ALL_USED_IDs <- COUNT_ID
COUNT_OBJ <- syn_temp$get(COUNT_ID, version = 1)
#COUNT <- read.table(COUNT_OBJ$path, header=T, sep='\t', check.names = F, row.names = 1)
COUNT <- as.data.frame(data.table::fread(COUNT_OBJ$path, header=T, sep='\t', check.names = F))#, row.names = 1)
row.names(COUNT) <- as.character(COUNT$feature)
COUNT <- COUNT[ ,(colnames(COUNT) %in% 'feature')==F ]
COUNT <- as.matrix(COUNT)

#Toss Zero variance genes post residual correct
COUNT <- COUNT[ (row.names(COUNT) %in% c("ENSG00000239467", "ENSG00000005486", "ENSG00000160972", "ENSG00000165886", "ENSG00000002330", "ENSG00000149823", "ENSG00000139722", "ENSG00000169217", "ENSG00000127774", "ENSG00000122490", "ENSG00000130881", "ENSG00000126005", "ENSG00000171222", "ENSG00000070413", "ENSG00000211459")) == F, ]

Provenance <- 'syn21075913'
Prov <- syn_temp$get(Provenance, version = 1)
ALL_USED_IDs[length(ALL_USED_IDs)+1] = Provenance
Translator <- read.csv(Prov$path, header=T, row.names = 1)
#colnames(COUNT) <- as.character( Translator[ colnames(COUNT), ][,2] )


#COUNT <- COUNT[ row.names(COUNT)[ (row.names(COUNT) %in% c('N_unmapped', 'N_multimapping', 'N_noFeature', 'N_ambiguous'))==F] , ]
#COUNT[,grep('150_120419', colnames(COUNT))[2]] = NULL

#Need to filter low counts here to remove the PAR_Y Regions
COUNT <- COUNT[ rowSums(COUNT[,] > 0) / (ncol(COUNT)) >= 0.05, ]

# Convert rownames of counts from tracking id to ensemble gene id
tmp = data.frame(Gene.ID = rownames(COUNT)) %>%
  dplyr::mutate(ID = Gene.ID) %>%
  tidyr::separate(ID, c('ensembl_gene_id', 'position'), sep = '\\.')
rownames(tmp) = tmp$Gene.ID
rownames(COUNT) = tmp[rownames(COUNT), 'ensembl_gene_id']

# Download expression data
Meta_ID <- 'syn23573928';
ALL_USED_IDs <- c( Meta_ID, ALL_USED_IDs )
Meta_OBJ <- syn_temp$get( Meta_ID, version = 2 )
Meta <- read.csv( Meta_OBJ$path, header=T, check.names = F, row.names = 1 )

#Filter for Rin Here#
Meta <- Meta[ Meta$RIN >= 5 | Meta$sampleid %in% c('380_120503','500_120515'), ]

#Filter Out Sample Swaps
Meta <- Meta[ Meta$FailQC_SampleSwapped == 0, ]
COUNT <- COUNT[ ,row.names(Meta) ]

 # table( Meta$tissue, Meta$study )      
 #       MAP ROS
 # ACC   387 338
 # DLPFC 639 463
 # PCC   388 265

METADATA <- Meta[ ,c( "individualID", "apoe4", "batch", "notes", "tissue", "diagnosis", "race", "pmi", "sex", "race", "spanish", "braaksc", "ceradsc", "cogdx", "dcfdx_lv", "pmi", "rincontinuous", "rin2", "age_death", "pct_pf_reads_aligned", "pct_coding_bases", "pct_intergenic_bases", "pct_intronic_bases" ) ]

GENE.PARAM <- syn_temp$get('syn23625835', version = 4)$path %>% 
  data.table::fread(data.table = T)
ALL_USED_IDs = c(ALL_USED_IDs, 'syn23625835')
GENE.LEN = dplyr::select(GENE.PARAM, ensembl_gene_id, gene_length) %>% 
  unique() 
rownames(GENE.LEN) = GENE.LEN$ensembl_gene_id
GENE.GC = dplyr::select(GENE.PARAM, ensembl_gene_id, percentage_gene_gc_content) %>% 
  unique() 
rownames(GENE.GC) = GENE.GC$ensembl_gene_id 

#Might need to rename
#percentage_gc_content - gene.length
```

### Data preprocessing
```{r preprocess.data, cache=TRUE}
# Remove samples with no cogdx, RIN, PMI scores and age_death
METADATA <- METADATA[, (colnames(METADATA)%in%'pmi.1')==F]

colnames(METADATA)[ colnames(METADATA) == 'race.1' ] <- 'race'
colnames(METADATA)[ colnames(METADATA) == 'rincontinuous' ] <- 'rin'

METADATA <- as.data.frame( cbind( Sampleid = row.names(METADATA), METADATA[ ,colnames(METADATA) ] ))
  
METADATA <- METADATA %>%
  ungroup %>%
  dplyr::filter(Sampleid %in% colnames(COUNT)) %>%
  dplyr::filter(!is.na(cogdx), !is.na(braaksc), !is.na(ceradsc)) %>%
  dplyr::filter(!is.na(rin)) %>%
  dplyr::filter(!is.na(pct_intronic_bases)) %>%
  dplyr::filter(!is.na(pmi)) %>%
  dplyr::filter(!is.na(age_death)) %>%
  as.data.frame()

# Add harmonised case-control status
METADATA$cogdx <- as.numeric(METADATA$cogdx)
METADATA$braaksc <- as.numeric(METADATA$braaksc)
METADATA$ceradsc <- as.numeric(METADATA$ceradsc)

METADATA$Diagnosis = 'OTHER'
METADATA$Diagnosis[METADATA$cogdx == 1 & METADATA$braaksc <= 3 & METADATA$ceradsc >= 3] = 'CONTROL'
METADATA$Diagnosis[METADATA$cogdx == 4 & METADATA$braaksc >= 4 & METADATA$ceradsc <= 2] = 'AD'

METADATA <- METADATA[ , (colnames(METADATA) %in% "diagnosis") == F ]

```
Dorsolateral prefrontal cortex of `r dim(COUNT)[2]` subjects from the ROS and MAP cohorts are used for the analysis. Following sample are removed due to missing metadata 

### Covariate clustering
Determine relationship between covariates
```{r covariates.clustering, cache=FALSE}
METADATA$Tissue.Diagnosis = paste(METADATA$tissue, METADATA$Diagnosis, sep = '.')
METADATA$Tissue.APOE4 = paste(METADATA$tissue, METADATA$apoe4, sep = '.')

primaryVariable <- c("cogdx", "Tissue.Diagnosis", "Tissue.APOE4", "dcfdx_lv", "braaksc", "ceradsc" )

#"Diagnosis", "apoe4",
FactorCovariates <- c( "Sampleid", "Tissue.Diagnosis", "Tissue.APOE4", "individualID", "Diagnosis", "batch", "sex", "race", "spanish", "cogdx", "cogdx", "ceradsc", "braaksc", "notes", "tissue", "dcfdx_lv" )

ContCovariates <- c( "rin", "rin2", "age_death", "pmi", "pct_pf_reads_aligned", 
                     "pct_coding_bases", "pct_intronic_bases", "pct_intergenic_bases" )

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
COVARIATES <- data.frame(lapply(COVARIATES,function(x){x <- sapply(x,function(y){str_replace_all(as.character(y),'\\+','')})}))
rownames(COVARIATES) <- METADATA$Sampleid
# Convert factor covariates to factors
#_#COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
if('cogdx.1' %in% colnames(COVARIATES)){
  COVARIATES <- COVARIATES[ ,(colnames(COVARIATES) %in% 'cogdx.1')==F ]
}
COVARIATES <- COVARIATES[ ,(colnames(COVARIATES) %in% 'Sampleid')==F ]
```

Correlation/association between covariates at an FDR <= 0.1
```{r covariates.correlation, cache=FALSE, fig.width=9, fig.height=6}
COVARIATES.CORRELATION = CovariateAnalysis::getAssociationStatistics(COVARIATES, PVAL = 0.05)
draw(COVARIATES.CORRELATION$plot, heatmap_legend_side = 'left', padding  = unit(c(18,2,2,18), 'mm')) 
```

### Explore metadata
```{r data.explore, cache=FALSE, fig.width = 10, fig.height = 14}
my.theme = theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1))
# RIN
p = list()
p[[1]] = ggplot(COVARIATES, aes(x = Tissue.Diagnosis, y = rin)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + my.theme
# Age Of Death
p[[2]] = ggplot(COVARIATES, aes(x = Tissue.Diagnosis, y = as.numeric(age_death))) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('AOD') + my.theme
# PMI
p[[3]] = ggplot(COVARIATES, aes(x = Tissue.Diagnosis, y = as.numeric(pmi))) + geom_boxplot()
p[[3]] = p[[3]] + ggtitle('PMI') + my.theme
# Ribosomal bases
p[[4]] = ggplot(COVARIATES, aes(x = Tissue.Diagnosis, y = as.numeric(pct_pf_reads_aligned))) + geom_boxplot()
p[[4]] = p[[4]] + ggtitle('Percent of pf aligned reads Bases') + my.theme
# Intronic bases
p[[5]] = ggplot(COVARIATES, aes(x = Tissue.Diagnosis, y = as.numeric(pct_intronic_bases))) + geom_boxplot()
p[[5]] = p[[5]] + ggtitle('Fraction Intronic Bases') + my.theme
# Intergenic bases
p[[6]] = ggplot(COVARIATES, aes(x = Tissue.Diagnosis, y = as.numeric(pct_intergenic_bases))) + geom_boxplot()
p[[6]] = p[[6]] + ggtitle('Fraction Intergenic Bases') + my.theme
multiplot(plotlist = p, cols = 3)
```

### Filter genes
* Remove genes that have less than 1 cpm counts in at least 50% of samples per Diagnosis
* Remove genes with missing gene length and percentage GC content
```{r filter.genes}
genesToAnalyze = COVARIATES %>%
  rownameToFirstColumn('SampleID') %>%
  dlply(.(Tissue.Diagnosis), .fun = function(mtd, count){
    processed.counts = getGeneFilteredGeneExprMatrix(count[,mtd$SampleID],
                                                     MIN_GENE_CPM=1, 
                                                     MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
    processed.counts$filteredExprMatrix$genes
  }, COUNT)
#filter out non-variable genes post residual removal

genesToAnalyze = unlist(genesToAnalyze) %>% 
  unique() %>%
  intersect(GENE.GC$ensembl_gene_id[!is.na(GENE.GC$percentage_gene_gc_content)]) %>%
  intersect(GENE.LEN$ensembl_gene_id[!is.na(GENE.LEN$gene_length)]) %>%
  setdiff(c("N_unmapped", "N_multimapping", "N_noFeature", "N_ambiguous")) %>%
  setdiff( c("ENSG00000239467", "ENSG00000005486", "ENSG00000160972", "ENSG00000165886", "ENSG00000002330", "ENSG00000149823", "ENSG00000139722", "ENSG00000169217", "ENSG00000127774", "ENSG00000122490", "ENSG00000130881", "ENSG00000126005", "ENSG00000171222", "ENSG00000070413", "ENSG00000211459") )

#18404
#vs
#18415
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[genesToAnalyze, ], 
                                                 MIN_GENE_CPM=0, 
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)

Ensemble2HGNC <- GENE.PARAM[ GENE.PARAM$ensembl_gene_id %in% PROCESSED_COUNTS$filteredExprMatrix$genes$genes, ] 
                      
summary(factor(Ensemble2HGNC$gene_biotype)) %>%
  rownameToFirstColumn('Biotype') %>%
  dplyr::rename(fraction = DF) %>%
  dplyr::mutate(fraction = fraction/dim(PROCESSED_COUNTS$filteredExprMatrix$genes)[1]) %>%
  dplyr::filter(fraction >= 0.01) %>%
  kable
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples

### Library Normalisation
Library normalisation is performed using cqn (conditional quantile normalisation)
```{r cqn}
GENE.GC <- as.data.frame(GENE.GC)
row.names(GENE.GC) <- GENE.GC$ensembl_gene_id
GENE.LEN <- as.data.frame(GENE.LEN)
row.names(GENE.LEN) <- GENE.LEN$ensembl_gene_id
# Compute offset for gene length and gc content
CQN.GENE_EXPRESSION = cqn(PROCESSED_COUNTS$filteredExprMatrix$counts, 
                          x = GENE.GC[ PROCESSED_COUNTS$filteredExprMatrix$genes$genes, 'percentage_gene_gc_content'],#$percentage_gene_gc_content,
                          lengths = GENE.LEN[PROCESSED_COUNTS$filteredExprMatrix$genes$genes, 'gene_length'],
                          lengthMethod = "smooth", 
                          verbose = FALSE)
CQN.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$y + CQN.GENE_EXPRESSION$offset
```

### Outlier Analysis
#### Sample outliers
Outlier analysis is performed before library normalisation with raw cpm counts
```{r outlier.analysis, cache = FALSE, fig.width = 10}
indToRemove = c('380_120503', '500_120515')
# Find principal components of expression to plot
PC <- prcomp(CQN.GENE_EXPRESSION$E, scale.=T, center = T)
# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])
plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  dplyr::mutate(label = SampleID) %>%
  tidyr::separate(Tissue.Diagnosis, c('Tissue', 'Diagnosis'), sep = '\\.')
plotdata$label[!(plotdata$SampleID %in% indToRemove)] = ''
p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(shape=notes, color=batch, size=rin))
p <- p + theme_bw() + theme(legend.position="top") 
p <- p + facet_grid(Tissue~.+Diagnosis, scales = 'free')
p <- p + geom_text(aes(label= label), size=4, hjust=0)
p

# Plot abberent distribution of logcpm counts
tmp1 = CQN.GENE_EXPRESSION$E %>%
  rownameToFirstColumn('Gene.ID') %>%
  tidyr::gather(SampleID, logCPM, -Gene.ID) %>%
  dplyr::left_join(COVARIATES %>%
                     rownameToFirstColumn('SampleID'))
p = ggplot(tmp1 %>%
             dplyr::filter(SampleID %in% indToRemove),
           aes(x = logCPM, color = SampleID)) + geom_density() 
p = p + theme(legend.position = 'top')
p
indToRetain = setdiff(colnames(PROCESSED_COUNTS$filteredExprMatrix$counts), indToRemove)
PROCESSED_COUNTS$filteredExprMatrix$counts = PROCESSED_COUNTS$filteredExprMatrix$counts[,indToRetain]
CQN.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$E[,indToRetain]
COVARIATES = COVARIATES[indToRetain,]
tmp = COVARIATES %>%
  tidyr::separate(Tissue.Diagnosis, c('Tissue', 'Diagnosis'), sep = '\\.') %>%
  dplyr::group_by(Diagnosis, Tissue) %>%
  dplyr::summarise(count = n()) %>%
  tidyr::spread(Diagnosis, count)
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples

Based on the expression pattern following samples were tagged as outliers: `r paste(indToRemove, collapse = ', ')`

Distribution of samples are: `r kable(tmp)`

#### Gene outliers
Assign NA values to genes that are above and below 3 std deviation of its distribution
```{r winsorise.data}
# Set gene counts in specific samples that are deviating 3 sd from other samples to 3SD limit
LOG.CPM = apply(CQN.GENE_EXPRESSION$E, 1, function(x){
  mn = mean(x, na.rm = T)
  std.dev = sd(x, na.rm = T)
  
  x[x < (mn-3*std.dev)] = NA
  x[x > (mn+3*std.dev)] = NA
  return(x)
}) %>% t
CQN.GENE_EXPRESSION$E = LOG.CPM
CQN.GENE_EXPRESSION$E.no.na = CQN.GENE_EXPRESSION$E
CQN.GENE_EXPRESSION$E.no.na[is.na(CQN.GENE_EXPRESSION$E.no.na)] = 0
LIB.SIZE = colSums(PROCESSED_COUNTS$filteredExprMatrix$counts)
NEW.COUNTS = (2^LOG.CPM) * t(replicate(dim(LOG.CPM)[1], LIB.SIZE))/1e6
```

### Sample clustering
PCA based clustering of samples
```{r decompse.normalise.data, cache=FALSE, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(CQN.GENE_EXPRESSION$E.no.na, scale.=T, center = T)
# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])
plotdata <- left_join(plotdata[row.names(COVARIATES),], rownameToFirstColumn(COVARIATES, 'SampleID'))
p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=batch, shape=notes, size=rin))
p <- p + theme_bw() + theme(legend.position="right")+ facet_grid(.~tissue, scales = 'free_y')
p
```

Tree based clustering of samples
```{r decompse.normalise.data.1, cache=FALSE, fig.height=6, fig.width=10, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,c("tissue", "notes","batch", "sex", "Tissue.Diagnosis")])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0
tree = hclust(as.dist(t(CQN.GENE_EXPRESSION$E.no.na)))
cols = WGCNA::labels2colors(COVARIATES.tmp);
WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```

### Distribution of samples (log cpm)
```{r lcpm.dist, cache=FALSE, fig.height=10, fig.width=20}
# Plot abberent distribution of logcpm counts
tmp1 = CQN.GENE_EXPRESSION$E %>%
  rownameToFirstColumn('Gene.ID') %>%
  tidyr::gather(SampleID, logCPM, -Gene.ID) %>%
  left_join(COVARIATES %>%
              tidyr::separate(Tissue.Diagnosis, c('Tissue', 'Diagnosis'), sep = '\\.') %>%
              rownameToFirstColumn('SampleID'))
p = ggplot(tmp1, aes(x = logCPM, color = SampleID)) + geom_density() 
p = p + theme(legend.position = 'NONE') + facet_grid(.~Diagnosis, scale = 'free')
p
```

Coexpression of genes 
```{r coexp1, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(CQN.GENE_EXPRESSION$E.no.na))
Temp_hist <- hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates are used to find significant covariates
```{r preadj.covariates, cache=FALSE}
COVARIATES <- COVARIATES[, (colnames(COVARIATES) %in% c("tissue"))==F] %>% 
  tidyr::separate(Tissue.APOE4, c('Tissue', 'apoe4'), sep = '\\.') %>%
  tidyr::separate(Tissue.Diagnosis, c('tissue', 'Diagnosis'), sep = '\\.')
COVARIATES$Tissue.APOE4 <- as.factor( paste0(as.character(COVARIATES$tissue), 
                                             '.', 
                                             as.character(COVARIATES$apoe4) ) )
COVARIATES$Tissue.Diagnosis <- as.factor( paste0(as.character(COVARIATES$tissue) ,
                                                 '.', 
                                                 as.character(COVARIATES$Diagnosis) ) )
# Find correlation between PC's of gene expression with covariates
COVARIATES <- COVARIATES[,  c( "individualID", "tissue", "Diagnosis", 'Tissue.Diagnosis', 'Tissue.APOE4',
                                                     "race", "spanish", "braaksc", "ceradsc",
                                                     "cogdx", "dcfdx_lv", "apoe4", "sex", "batch",
                                                     "notes", "pmi", "rin", "rin2", "age_death",
                                                     "pct_pf_reads_aligned","pct_pf_reads_aligned",
                                                     "pct_intronic_bases", "pct_intronic_bases",
                               "pct_intergenic_bases", "pct_coding_bases"
                                                              )]
COVARIATES$tissue <- as.factor(COVARIATES$tissue)
COVARIATES$Diagnosis <- as.factor(COVARIATES$Diagnosis)
COVARIATES$apoe4 <- as.factor(COVARIATES$apoe4)

COVARIATES <- COVARIATES[colnames(CQN.GENE_EXPRESSION$E.no.na),]

COVARIATES <- COVARIATES[ , colnames(COVARIATES)[grepl('[.]1',colnames(COVARIATES))==F ]]

preAdjustedSigCovars = runPCAandPlotCorrelations(CQN.GENE_EXPRESSION$E.no.na, 
                                                 COVARIATES,
                                                 'NULL design(voom-normalized)', 
                                                 isKeyPlot=TRUE, 
                                                 MIN_PVE_PCT_PC = 1)
```

Significant covariates to adjust at FDR 0.1 are `r preAdjustedSigCovars$significantCovars`
```{r preadj.covariates.plot, cache=FALSE, fig.width=12, fig.height=8}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

### Normalisation (iterative design)
Since many covariates are correlated, re-normalising and re-adjusting COUNTS with an iterative design matrix
1. Adding Batch and Sex a priori to variable selection
2. Primary variable of interest Diagnosis is excluded from the pool of available covariates for selection
```{r iterative.norm, cache=FALSE, results='asis'}
# Primary variable of interest
postAdjustCovars = c('batch', 'sex');
# Assign residual covariates
residualCovars = setdiff(preAdjustedSigCovars$significantCovars, c(postAdjustCovars, primaryVariable, 'individualID', 'tissue' ))
residualSigCovars = preAdjustedSigCovars
covariatesEffects = preAdjustedSigCovars$Effects.significantCovars[residualCovars]

#_# ADD DIAGNOSIS HERE TO REGRESS FOR TWAS
postAdjustCovars = c(postAdjustCovars,"Diagnosis", names(which.max(covariatesEffects))) %>% unique()
loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 20){
  writeLines(paste('Using following covariates in the model:',
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
  
  # Estimate voom weights for dispersion control
  cnts = NEW.COUNTS
  cnts[is.na(cnts)] = 0
  VOOM.GENE_EXPRESSION = voom(cnts, 
                              design=DM1$design, 
                              plot=F,
                              na.rm = T)
  
  # Fit linear model using new weights and new design
  VOOM.ADJUSTED.FIT = lmFit(CQN.GENE_EXPRESSION$E,
                            design = DM1$design,
                            weights = VOOM.GENE_EXPRESSION$weights)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,
                                                CQN.GENE_EXPRESSION$E)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), c(postAdjustCovars, primaryVariable, "individualID", "Sampleid"))
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  expr = RESIDUAL.GENE_EXPRESSION
  expr[is.na(expr)] = 0
  residualSigCovars = runPCAandPlotCorrelations(expr, 
                                                COVARIATES[, residCovars, drop=F], 
                                                'adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  residCovars = setdiff(residualSigCovars$significantCovars, c(postAdjustCovars, primaryVariable, "tissue"))
  covariatesEffects = residualSigCovars$Effects.significantCovars[residCovars]
  
  postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}
modelStr <- paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=', '),
                  'as fixed effects')
tmp <- paste('Using following covariates in the final model:', modelStr)
```
`r tmp`

### Sanity check
```{r residual.adj, cache=FALSE, fig.width=12, fig.height=8}
# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
#COVARIATES$tissue <- as.factor(COVARIATES$tissue)
#COVARIATES$Diagnosis <- as.factor(COVARIATES$Diagnosis)
#COVARIATES$Tissue <- as.factor(COVARIATES$Tissue)
#COVARIATES$apoe4 <- as.factor(COVARIATES$apoe4)


#COVARIATES <- COVARIATES[ ,(colnames(COVARIATES) %in% 'pct_intronic_bases.1')==F ]
residualSigCovars = runPCAandPlotCorrelations(expr, 
                                              COVARIATES[colnames(expr),],
                                              'adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)
residualSigCovars[["PC_res"]][[2]]$plotData
```
Coexpression of genes 
```{r coexp2, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(expr))
Temp_Histb <- hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```
PCA of residual data
```{r decompse.normalise.data2.1, cache=FALSE, fig.height=6, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(expr, scale.=T, center = T)
# Plot first 4 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])
plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID'))
p <- ggplot(plotdata, aes(x=PC1, y=PC2)) 
p <- p + geom_point(aes(color=batch, shape=notes, size=rin))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(.~tissue, scales = 'free_y')
p
```
Tree based clustering of residual data
```{r decompse.normalise.data2.2, cache=FALSE, fig.height=6, fig.width=10, results='asis'}
# Eucledian tree based analysis
COVARIATES$Tissue.Diagnosis <- as.factor( paste0( COVARIATES$tissue, '.', COVARIATES$Diagnosis) )
COVARIATES$Tissue.APOE4 <-as.factor( paste0( COVARIATES$tissue, '.', COVARIATES$apoe4) )
COVARIATES.tmp = data.matrix(COVARIATES[,c('batch', 'tissue', 'apoe4', 'Diagnosis','sex', primaryVariable)])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0
tree = hclust(as.dist(t(expr)))
cols = WGCNA::labels2colors(COVARIATES.tmp);
WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```

### Adjust data with covariates for Network Analysis
Identified covariates are regressed out from the expression matrix for network analysis
```{r network.adjust}
# Get design matrix
DESIGN.NET = getDesignMatrix(COVARIATES[, postAdjustCovars, drop = F], Intercept = F)
DESIGN.NET = DESIGN.NET$design[,linColumnFinder(DESIGN.NET$design)$indepCols]
# Estimate voom weights for dispersion control
cnts = NEW.COUNTS
cnts[is.na(cnts)] = 0
VOOM.NET.WEIGHTS = voom(cnts, design=DESIGN.NET, plot=F)
# Fit linear model using new weights and new design
VOOM.NET.FIT = lmFit(CQN.GENE_EXPRESSION$E,
                     design = DESIGN.NET,
                     weights = VOOM.NET.WEIGHTS$weights)
# Residuals after normalisation
RESIDUAL.NET.GENE_EXPRESSION = residuals.MArrayLM(VOOM.NET.FIT,
                                                  CQN.GENE_EXPRESSION$E)
```

### SVA Adjustments for eQTL analysis
Conditioned on primary variable (Tissue.Diagnosis) and identified covariates estimate surrogate variables using SVA package
```{r sva.adjust, cache=FALSE }
# Get design matrix
DESIGN = getDesignMatrix(COVARIATES[, c(primaryVariable[2], postAdjustCovars), drop = F], Intercept = F)
DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
# Estimate voom weights for dispersion control
cnts = NEW.COUNTS
cnts[is.na(cnts)] = 0
VOOM.WEIGHTS = voom(cnts, design=DESIGN$design, plot=F)
# Fit linear model using new weights and new design
FIT = lmFit(CQN.GENE_EXPRESSION$E,
            design = DESIGN$design,
            weights = VOOM.WEIGHTS$weights)
# Get (null) design matrix
MODEL0 = DESIGN$design[,-grep(primaryVariable[2],colnames(DESIGN$design))] # Get null model by removing variable of interest
MODEL0 = MODEL0[,linColumnFinder(MODEL0)$indepCols]
MODEL1 = DESIGN$design
MODEL1 = MODEL1[,linColumnFinder(MODEL1)$indepCols]
# Calculate the number of surrogate variables (be method)
# NUM.SV = sva::num.sv(CQN.GENE_EXPRESSION$E, MODEL1, method = 'be', seed = 123456, B = 30)
# Get residuals from differential expression model
RESIDUAL.SVA.GENE_EXPRESSION = residuals.MArrayLM(FIT, CQN.GENE_EXPRESSION$E)
# Compute actual variance of all principal components
expr = RESIDUAL.SVA.GENE_EXPRESSION
expr[is.na(expr)] = 0
tmp = svd(expr)
actual.var = tmp$d^2/sum(tmp$d^2)
# Compute permuted variance of all principal components
permuted.var = foreach(i = 1:20, .combine = rbind) %dopar% {
  tmp.residual = t(apply(expr, 1, sample, replace = FALSE))
  tmp = svd(tmp.residual)
  permuted.var = tmp$d^2/sum(tmp$d^2)
}
permuted.var = apply(permuted.var, 2, mean)
NUM.SV = sum(actual.var >= permuted.var[1])
# Plot variance components
var.comp = data.frame(component = 1:length(actual.var), 
                      residual = actual.var, 
                      permuted.residual = permuted.var) %>%
  tidyr::gather(data, value, -component)
p = ggplot(var.comp %>% filter(component <= round(NUM.SV*1.3)), aes(x = component, y = value, color = data)) + geom_point()
p = p + geom_hline(yintercept=permuted.var[1], linetype = 'dashed') + xlab('Singular Dimension Index')
p = p + geom_vline(xintercept=NUM.SV, linetype = 'dashed') + ylab('Fraction of Variance Explained')
p = p + theme(legend.position = c(0.8, 0.8), legend.title = element_blank())
p

# Estimate surrogate variables
SURR.VAR = sva::sva(CQN.GENE_EXPRESSION$E.no.na, 
                    MODEL1, MODEL0, n.sv = NUM.SV, B = 30)$sv
SURR.VAR = data.frame(SURR.VAR)
colnames(SURR.VAR) = paste0('SV',1:dim(SURR.VAR)[2])
rownames(SURR.VAR) = rownames(MODEL1)
# Re-estimate voom weights
VOOM.SVA.WEIGHTS = voom(cnts, 
                        design = cbind(MODEL1, SURR.VAR), plot=F)
# Fit linear model using new weights and new design
VOOM.SVA.FIT = lmFit(CQN.GENE_EXPRESSION$E, 
                     design = cbind(MODEL1, SURR.VAR),
                     weights = VOOM.SVA.WEIGHTS$weights)
# Residuals after normalisation
RESIDUAL.SVA.GENE_EXPRESSION = residuals.MArrayLM(VOOM.SVA.FIT,
                                                  CQN.GENE_EXPRESSION$E)
# Add variable of interest back to the residuals
varsToAddIn = grep("Diagnosis", colnames(MODEL1), value = T)
RESIDUAL.SVA.GENE_EXPRESSION = RESIDUAL.SVA.GENE_EXPRESSION + 
  VOOM.SVA.FIT$coefficients[,varsToAddIn] %*% t(VOOM.SVA.FIT$design[,varsToAddIn])
```


### Differential expression analysis (with cogdx as primary variable)
Differential expression is performed on the primary variable by controlling for covariates identified above

Interpretation of cogdx scores
1. NCI, No cognitive impairment (No impaired domains)
2. MCI, Mild cognitive impairment (One impaired domain) and NO other cause of CI
3. MCI, Mild cognitive impairment (One impaired domain) AND another cause of CI
4. AD, Alzheimer's disease and NO other cause of CI (NINCDS PROB AD)
5. AD, Alzheimer's disease AND another cause of CI (NINCDS POSS AD)
6. Other dementia. Other primary cause of dementia

Genes that are differentially expressed at an FDR <= 0.05 are
```{r diff.exp, fig.height=6, fig.width=18, cache=FALSE}

COVARIATES.tmp = COVARIATES
COVARIATES.tmp$Tissue.cogdx <- as.factor(paste0( COVARIATES.tmp$tissue, '.', COVARIATES.tmp$cogdx ))
"tissue"
DESIGN = getDesignMatrix( COVARIATES.tmp[, c("Tissue.cogdx", postAdjustCovars[ (postAdjustCovars %in% c( "dcfdx_lv","braaksc","ceradsc" ))==F ]), drop = F], Intercept = F)
ind = grep('tissue', colnames(DESIGN$design))

DESIGN$design[,ind] = DESIGN$design[,ind] * as.numeric(COVARIATES.tmp[,'cogdx'])
colnames(DESIGN$design)[ind] = paste0(colnames(DESIGN$design),'cogdx')
DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
# Re-estimate voom weights
VOOM.WEIGHTS = voom(cnts,
                    design=DESIGN$design, plot=F,
                    block = COVARIATES$individualIdentifier, 
                    correlation = correlation$cor)

VOOM.WEIGHTS$E = CQN.GENE_EXPRESSION$E
FIT = lmFit(VOOM.WEIGHTS)
# Fit contrast
contrast = makeContrasts(contrasts = c("Tissue.cogdxACC.2-Tissue.cogdxACC.1",
                                       "Tissue.cogdxACC.4-Tissue.cogdxACC.1",
                                       "Tissue.cogdxACC.4-Tissue.cogdxACC.2",
                                       "Tissue.cogdxDLPFC.2-Tissue.cogdxDLPFC.1",
                                       "Tissue.cogdxDLPFC.4-Tissue.cogdxDLPFC.1",
                                       "Tissue.cogdxDLPFC.4-Tissue.cogdxDLPFC.2",
                                       "Tissue.cogdxPCC.2-Tissue.cogdxPCC.1",
                                       "Tissue.cogdxPCC.4-Tissue.cogdxPCC.1",
                                       "Tissue.cogdxPCC.4-Tissue.cogdxPCC.2"),
                         levels = colnames(FIT$coefficients))
FIT.CONTR = contrasts.fit(FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)
# Get differnetial expression
DE = lapply(1:9, function(i, FIT){
  topTable(FIT, coef=i, number = 50000, confint = T) %>%
    rownameToFirstColumn('ensembl_gene_id')
}, FIT.CONTR)
names(DE) = gsub("Tissue.","", colnames(contrast))
DE = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Tissue','',Comparison),
                Study = 'ROSMAP',
                Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>%
  dplyr::rename(Tissue = Comparison) %>%
  dplyr::mutate(Comparison = Tissue) %>%
  left_join(GENE.PARAM %>%
              dplyr::select(ensembl_gene_id, hgnc_symbol, percentage_gene_gc_content, gene_length) %>%
              unique())

DE$Tissue <- gsub( 'cogdx','', do.call(rbind, strsplit( do.call(rbind,strsplit(DE$Tissue,'-'))[,2],
                                      '[.]'))[,1])
DE$Comparison <- gsub( 'ACC.','', DE$Comparison)
DE$Comparison <- gsub( 'PCC.','', DE$Comparison)
DE$Comparison <- gsub( 'DLPFC.','', DE$Comparison)

DE$Direction[DE$adj.P.Val > 0.05 | abs(DE$logFC) < log2(1.2)] = 'NONE'
tmp = DE %>%
  dplyr::select(ensembl_gene_id, Comparison, Direction, Tissue) %>%
  group_by(Tissue, Comparison, Direction) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id))) %>%
  spread(Direction, FDR_0_05_FC_1.2) 
kable(tmp)
p = ggplot(DE, aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red')) + theme(legend.position = 'top')
p = p + facet_grid(Comparison+.~Tissue, scales = 'free')
p
all.diff.exp = list(cogdx = DE)
all.fit = list(cogdx = FIT)
```

### Differential expression analysis (with Diagnosis as primary variable)
Differential expression is performed on the primary variable by controlling for covariates identified above

Interpretation of Diagnosis
1. AD: cogdx = 4, braaksc >= 4 and ceradsc <= 2
2. CONTROL: cogdx = 1, braaksc <= 3 and ceradsc >= 3
3. OTHER: All the other samples

Genes that are differentially expressed at an FDR <= 0.05 and folde change of 1.2 are
```{r diff.exp1, fig.height=6, fig.width=18, cache=FALSE}
DESIGN = getDesignMatrix(COVARIATES[, c(primaryVariable[2], postAdjustCovars[ (postAdjustCovars %in% c( "dcfdx_lv","braaksc","ceradsc" ))==F ]), drop = F], Intercept = F)
DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
# Re-estimate voom weights
VOOM.WEIGHTS = voom(cnts,
                    design=DESIGN$design, plot=F,
                    block = COVARIATES$individualIdentifier, 
                    correlation = correlation$cor)
# Fit linear model using new weights and new design
VOOM.WEIGHTS$E = CQN.GENE_EXPRESSION$E
FIT = lmFit(VOOM.WEIGHTS)
# Fit contrast
cntr = c()
for (region in c('ACC','PCC','DLPFC')){
  tmp = combn(paste0('Tissue.Diagnosis', region, '.', c('AD', 'OTHER', 'CONTROL')),2)
  cntr = c(cntr, apply(tmp, 2, paste, collapse = '-'))
}
contrast = makeContrasts(contrasts=cntr,
                         levels = colnames(FIT$coefficients))
FIT.CONTR = contrasts.fit(FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)
# Get differnetial expression
DE = lapply(1:9, function(i, FIT){
  topTable(FIT, coef=i, number = 50000, confint = T) %>%
    rownameToFirstColumn('ensembl_gene_id')
}, FIT.CONTR)
names(DE) = colnames(contrast)
DE = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Tissue.Diagnosis','',Comparison),
                Study = 'ROSMAP',
                Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>%
  tidyr::separate(Comparison, c('from.state','to.state'), sep = '-') %>%
  tidyr::separate(from.state, c('Tissue','reference'), sep = '\\.') %>%
  tidyr::separate(to.state, c('Tissue1','against'), sep = '\\.') %>%
  dplyr::mutate(Comparison = paste(reference, against, sep = '-')) %>%
  dplyr::select(-Tissue1) %>%
  left_join(GENE.PARAM %>%
              dplyr::select(ensembl_gene_id, hgnc_symbol, percentage_gene_gc_content, gene_length) %>%
              unique())
DE$Direction[DE$adj.P.Val > 0.05 | abs(DE$logFC) < log2(1.2)] = 'NONE'
tmp = DE %>%
  dplyr::select(ensembl_gene_id, Comparison, Direction, Tissue) %>%
  group_by(Tissue, Comparison, Direction) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id))) %>%
  spread(Direction, FDR_0_05_FC_1.2) 
kable(tmp)
p = ggplot(DE, aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red')) + theme(legend.position='top')
p = p + facet_grid(Tissue+.~Comparison, scales = 'fixed')
p
all.diff.exp = c(all.diff.exp, list(Diagnosis = DE))
all.fit = c(all.fit, list(Diagnosis = FIT))
```

### Associate differential expression results with gc content, gene length and average expression
```{r associate.de, fig.height=10, fig.width=15}
pl = list()
pl[[1]] = ggplot(DE %>% dplyr::filter(Comparison == 'AD-CONTROL'), 
                 aes(x = log10(gene_length), y = logFC, color = Direction)) + geom_point() 
pl[[1]] = pl[[1]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = log10(gene_length), y = logFC))
pl[[1]] = pl[[1]] + facet_grid(Tissue~.+Comparison) + scale_color_manual(values = c('green','grey','red'))
pl[[1]] = pl[[1]] + theme(legend.position = 'top')
pl[[2]] = ggplot(DE %>% dplyr::filter(Comparison == 'AD-CONTROL'), 
                 aes(x = percentage_gene_gc_content, y = logFC, color = Direction)) + geom_point()
pl[[2]] = pl[[2]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = percentage_gene_gc_content, y = logFC))
pl[[2]] = pl[[2]] + facet_grid(Tissue~.+Comparison) + scale_color_manual(values = c('green','grey','red'))
pl[[2]] = pl[[2]] + theme(legend.position = 'top')
pl[[3]] = ggplot(DE %>% dplyr::filter(Comparison == 'AD-CONTROL'), 
                 aes(x = AveExpr, y = logFC, color = Direction)) + geom_point()
pl[[3]] = pl[[3]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = AveExpr, y = logFC))
pl[[3]] = pl[[3]] + facet_grid(Tissue~.+Comparison) + scale_color_manual(values = c('green','grey','red'))
pl[[3]] = pl[[3]] + theme(legend.position = 'top')
multiplot(plotlist = pl, cols = 3)

```

### Differential expression analysis (with APOE status as primary variable)
Differential expression is performed on the primary variable by controlling for covariates identified above

Interpretation of apoe_genotype
1. 0: No apoe4 allele
2. 1: 1 apoe4 allele
3. 2: 2 apoe4 allele

Genes that are differentially expressed at an FDR <= 0.05 and folde change of 1.2 are
```{r diff.exp2, fig.height=6, fig.width=18, cache=FALSE}
DESIGN = getDesignMatrix(COVARIATES[, c(primaryVariable[3], postAdjustCovars[ (postAdjustCovars %in% c( "dcfdx_lv","braaksc","ceradsc" ))==F ]), drop = F], Intercept = F)
DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
# Re-estimate voom weights
VOOM.WEIGHTS = voom(cnts,
                    design=DESIGN$design, plot=F,
                    block = COVARIATES$individualIdentifier, 
                    correlation = correlation$cor)
# Fit linear model using new weights and new design
VOOM.WEIGHTS$E = CQN.GENE_EXPRESSION$E
FIT = lmFit(VOOM.WEIGHTS)
# Fit contrast
cntr = c()
for (region in c('ACC','PCC','DLPFC')){
  tmp = combn(paste0('Tissue.APOE4', region, '.', c('0', '1', '2')),2)
  cntr = c(cntr, apply(tmp, 2, paste, collapse = '-'))
}
contrast = makeContrasts(contrasts=cntr,
                         levels = colnames(FIT$coefficients))
FIT.CONTR = contrasts.fit(FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)
# Get differnetial expression
DE = lapply(1:dim(contrast)[2], function(i, FIT){
  topTable(FIT, coef=i, number = 50000, confint = T) %>%
    rownameToFirstColumn('ensembl_gene_id')
}, FIT.CONTR)
names(DE) = colnames(contrast)

DE = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Tissue.APOE4','',Comparison),
                Study = 'ROSMAP',
                Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>%
  tidyr::separate(Comparison, c('from.state','to.state'), sep = '-') %>%
  tidyr::separate(from.state, c('Tissue','reference'), sep = '\\.') %>%
  tidyr::separate(to.state, c('Tissue1','against'), sep = '\\.') %>%
  dplyr::mutate(Comparison = paste(reference, against, sep = '-')) %>%
  dplyr::select(-Tissue1) %>%
  left_join(GENE.PARAM %>%
              dplyr::select(ensembl_gene_id, hgnc_symbol, percentage_gene_gc_content, gene_length) %>%
              unique())
DE$Direction[DE$adj.P.Val > 0.05 | abs(DE$logFC) < log2(1.2)] = 'NONE'
tmp = DE %>%
  dplyr::select(ensembl_gene_id, Comparison, Direction, Tissue) %>%
  group_by(Tissue, Comparison, Direction) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id))) %>%
  spread(Direction, FDR_0_05_FC_1.2) 
kable(tmp)
p = ggplot(DE, aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red')) + theme(legend.position='top')
p = p + facet_grid(Tissue+.~Comparison, scales = 'fixed')
p
all.diff.exp = c(all.diff.exp, list(APOE4 = DE))
all.fit = c(all.fit, list(APOE4 = FIT))
```

### Differential expression analysis (with Diagnosis and Sex) 
Differential expression is performed on the Diagnosis x Sex variable by controlling for covariates identified above

Genes that are differentially expressed at an FDR <= 0.05 and folde change of 1.2 are
```{r diff.exp3, fig.height=6, fig.width=18, cache=FALSE}
## Modeling Diagnosis, msex and age_death conjointly
COVARIATES$Tissue.Diagnosis.SEX = paste(COVARIATES$Tissue.Diagnosis,COVARIATES$sex, sep = '.') %>% factor
# Get design matrix
DESIGN = getDesignMatrix(COVARIATES[, c('Tissue.Diagnosis.SEX', setdiff(postAdjustCovars, 'SEX')), drop = F], Intercept = F)
DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
# Re-estimate voom weights
VOOM.WEIGHTS = voom(cnts,
                    design=DESIGN$design, plot=F,
                    block = COVARIATES$individualIdentifier, 
                    correlation = correlation$cor)
# Fit linear model using new weights and new design
VOOM.WEIGHTS$E = CQN.GENE_EXPRESSION$E
FIT = lmFit(VOOM.WEIGHTS)

colnames(FIT$coefficients) <- gsub( 'Tissue.Diagnosis.SEX','',colnames(FIT$coefficients))
# Fit contrast
contrast = makeContrasts(contrasts=c("(ACC.AD.female+ACC.AD.male)/2-(ACC.CONTROL.female+ACC.CONTROL.male)/2",
                                     "(ACC.AD.female+ACC.CONTROL.female)/2-(ACC.AD.male+ACC.CONTROL.male)/2",
                                     "(ACC.AD.female-ACC.CONTROL.female)/2-(ACC.AD.male-ACC.CONTROL.male)/2",
                                     "ACC.AD.female-ACC.CONTROL.female",
                                     "ACC.AD.male-ACC.CONTROL.male",
                                     "(DLPFC.AD.female+DLPFC.AD.male)/2-(DLPFC.CONTROL.female+DLPFC.CONTROL.male)/2",
                                     "(DLPFC.AD.female+DLPFC.CONTROL.female)/2-(DLPFC.AD.male+DLPFC.CONTROL.male)/2",
                                     "(DLPFC.AD.female-DLPFC.CONTROL.female)/2-(DLPFC.AD.male-DLPFC.CONTROL.male)/2",
                                     "DLPFC.AD.female-DLPFC.CONTROL.female",
                                     "DLPFC.AD.male-DLPFC.CONTROL.male",
                                     "(PCC.AD.female+PCC.AD.male)/2-(PCC.CONTROL.female+PCC.CONTROL.male)/2",
                                     "(PCC.AD.female+PCC.CONTROL.female)/2-(PCC.AD.male+PCC.CONTROL.male)/2",
                                     "(PCC.AD.female-PCC.CONTROL.female)/2-(PCC.AD.male-PCC.CONTROL.male)/2",
                                     "PCC.AD.female-PCC.CONTROL.female",
                                     "PCC.AD.male-PCC.CONTROL.male"),
                         levels = colnames(FIT$coefficients))
colnames(contrast) = c('ACC.AD.MALE-ACC.CONTROL.MALE',
                       'ACC.AD.FEMALE-ACC.CONTROL.FEMALE',
                       'ACC.AD.ALL-ACC.CONTROL.ALL',
                       'ACC.ALL.FEMALE-ACC.ALL.MALE',
                       'ACC.ALL.ALL-ACC.ALL.ALL',
                       'DLPFC.AD.MALE-DLPFC.CONTROL.MALE',
                       'DLPFC.AD.FEMALE-DLPFC.CONTROL.FEMALE',
                       'DLPFC.AD.ALL-DLPFC.CONTROL.ALL',
                       'DLPFC.ALL.FEMALE-DLPFC.ALL.MALE',
                       'DLPFC.ALL.ALL-DLPFC.ALL.ALL',
                       'PCC.AD.MALE-PCC.CONTROL.MALE',
                       'PCC.AD.FEMALE-PCC.CONTROL.FEMALE',
                       'PCC.AD.ALL-PCC.CONTROL.ALL',
                       'PCC.ALL.FEMALE-PCC.ALL.MALE',
                       'PCC.ALL.ALL-PCC.ALL.ALL'
                       )
FIT.CONTR = contrasts.fit(FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)
# Get differnetial expression
DE = lapply(1:dim(contrast)[2], function(i, FIT){
  topTable(FIT, coef=i, number = 50000, confint = T) %>%
    rownameToFirstColumn('ensembl_gene_id')
}, FIT.CONTR)
names(DE) = colnames(contrast)
DE = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Tissue.Diagnosis','',Comparison),
                Study = 'ROSMAP',
                Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>%
  left_join(GENE.PARAM %>%
              dplyr::select(ensembl_gene_id, hgnc_symbol, percentage_gene_gc_content, gene_length) %>%
              unique())
DE$Direction[DE$adj.P.Val > 0.05 | abs(DE$logFC) < log2(1.2)] = 'NONE'
DE$Tissue <- do.call( rbind, strsplit( DE$Comparison, '[.]'))[,1]
DE$Comparison <- gsub('DLPFC.', '',gsub('ACC.', '', gsub('PCC.', '',  DE$Comparison )))
tmp = DE %>%
  dplyr::select(ensembl_gene_id, Tissue, Comparison, Direction) %>%
  group_by(Tissue, Comparison, Direction) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id))) %>%
  spread(Direction, FDR_0_05_FC_1.2) 
kable(tmp)
DE$Tissue <- do.call( rbind, strsplit( DE$Comparison, '[.]'))[,1]
all.diff.exp = c(all.diff.exp, list(Diagnosis.SEX = DE))
all.fit = c(all.fit, list(Diagnosis.SEX = FIT))
```

### Differential expression analysis (with Diagnosis and AOD)
Differential expression is performed on the Diagnosis x AOD variable by controlling for covariates identified above

Genes that are differentially expressed at an FDR <= 0.05 and folde change of 1.2 are
```{r diff.exp4, fig.height=6, fig.width=18, cache=FALSE}
# Get design matrix
DESIGN = getDesignMatrix(COVARIATES[, c(primaryVariable[2], setdiff(postAdjustCovars, 'age_death')), drop = F], Intercept = F)
ind = grep('Diagnosis', colnames(DESIGN$design))
DESIGN$design[,ind] = DESIGN$design[,ind] * COVARIATES[rownames(DESIGN$design),'age_death']
DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]

colnames(DESIGN$design)[ind[1:9]] = gsub('Tissue.Diagnosis', '', paste0(colnames(DESIGN$design)[ind[1:9]], '.age_death'))
# Re-estimate voom weights
VOOM.WEIGHTS = voom(cnts,
                    design=DESIGN$design, plot=F,
                    block = COVARIATES$individualIdentifier, 
                    correlation = correlation$cor)
# Fit linear model using new weights and new design
VOOM.WEIGHTS$E = CQN.GENE_EXPRESSION$E
FIT = lmFit(VOOM.WEIGHTS)
cntr = c("ACC.AD.age_death-ACC.CONTROL.age_death",
         "0.5*ACC.AD.age_death+0.5*ACC.CONTROL.age_death",
         "DLPFC.AD.age_death-DLPFC.CONTROL.age_death",
         "0.5*DLPFC.AD.age_death+0.5*DLPFC.CONTROL.age_death",
         "PCC.AD.age_death-PCC.CONTROL.age_death",
         "0.5*PCC.AD.age_death+0.5*PCC.CONTROL.age_death"
        )
         
contrast = makeContrasts(contrasts=cntr,
                         levels = colnames(FIT$coefficients))
colnames(contrast) = c("ACC.AD.AOD-ACC.CONTROL.AOD",
                       "ACC.ALL.AOD",
                       "DLPFC.AD.AOD-DLPFC.CONTROL.AOD",
                       "DLPFC.ALL.AOD",
                       "PCC.AD.AOD-PCC.CONTROL.AOD",
                       "PCC.ALL.AOD"
                      )
FIT.CONTR = contrasts.fit(FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)
# Get differnetial expression
DE = lapply(1:dim(contrast)[2], function(i, FIT){
  topTable(FIT, coef=i, number = 50000, confint = T) %>%
    rownameToFirstColumn('ensembl_gene_id')
}, FIT.CONTR)
names(DE) = colnames(contrast)
DE = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Tissue.Diagnosis','',Comparison),
                Study = 'ROSMAP',
                Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>%
  left_join(GENE.PARAM %>%
              dplyr::select(ensembl_gene_id, hgnc_symbol, percentage_gene_gc_content, gene_length) %>%
              unique())
DE$Direction[DE$adj.P.Val > 0.05 | abs(DE$logFC) < log2(1.2)] = 'NONE'
DE$Tissue <- do.call( rbind, strsplit( DE$Comparison, '[.]'))[,1]
DE$Comparison <- gsub('DLPFC.', '',gsub('ACC.', '', gsub('PCC.', '',  DE$Comparison )))
tmp = DE %>%
  dplyr::select(ensembl_gene_id, Tissue, Comparison, Direction) %>%
  group_by(Tissue, Comparison, Direction) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id))) %>%
  spread(Direction, FDR_0_05_FC_1.2) 
kable(tmp)
all.diff.exp = c(all.diff.exp, list(Diagnosis.AOD = DE))
all.fit = c(all.fit, list(Diagnosis.AOD = FIT))
```

```{r NeuroPath_DataClean, include=FALSE, cache=TRUE}
EXP <- t( RESIDUAL.SVA.GENE_EXPRESSION )
FullMeta <- COVARIATES
PartialMeta <- FullMeta[ ,c('tissue', 'ceradsc', 'braaksc', 'cogdx', 'dcfdx_lv' ) ]
table( row.names(PartialMeta) == row.names(EXP) )


#### Setup and Run Models

NeuroPath_Calc <- function( GN, Path, Exp, Tiss ){
  #'@GN a character ENSG Gene name eg 'ENSG00000227232'
  #'@Path the character string of Neuropath column to use for model eg. EITHER: 'ceradsc', 'braaksc', 'c', 'dcfdx_lv'
  #'@Exp
  #'@Tiss eg. 'ACC', 'PCC', 'DLPFC'
  
  
  Dat <- as.data.frame( cbind( PartialMeta[ PartialMeta$tissue == Tiss ,], 
                               Gene = scale( as.numeric(Exp[ (row.names(PartialMeta[ PartialMeta$tissue == Tiss ,])) ,GN]) ) ), stringsAsFactors = F )
  
  Dat$ceradsc <- as.factor(Dat$ceradsc)
  Dat$braaksc <- as.factor(Dat$braaksc)
  Dat$cogdx <- as.factor(Dat$cogdx)
  Dat$dcfdx_lv <- as.factor(Dat$dcfdx_lv)
  
  m <- eval(parse(text=paste0( 'MASS::polr(', Path, '~ Gene, data = Dat, Hess=TRUE)' )))
  ctable <- coef(summary(m))
  
  p <- pnorm(abs(ctable[, "t value"]), lower.tail = FALSE) * 2
  ctable <- cbind(ctable, "p value" = p)
  #ERROR is HERE:
  ci <- confint(m)
  
  names( GN ) <- 'Gene'
  PVal <- p['Gene']
  names( PVal ) <- 'PVal'
  OR <- exp(coef(m))
  names( OR ) <- 'OR'
  CI <- exp(cbind(OR = coef(m), ci))[,2]
  Tissue <- Tiss
  return( c( GN, Tissue, OR, CI, PVal ) )
}

#assign gne names to a variable to iterate over
temp<-colnames(EXP)
```
```{r NeuroPath_Associtation_1_1, include=FALSE, cache=TRUE}
#BRAAK
Path <- 'braaksc'
mark<-Sys.time()
BRAAK_ACC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'ACC' )}, error=function(e) NULL)
Sys.time() - mark
```
```{r NeuroPath_Associtation_1_2, include=FALSE, cache=TRUE}
mark<-Sys.time()
BRAAK_PCC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'PCC' )}, error=function(e) NULL)
Sys.time() - mark
```

```{r NeuroPath_Associtation_1_3, include=FALSE, cache=TRUE}
mark<-Sys.time()
BRAAK_DLPFC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'DLPFC' )}, error=function(e) NULL)
Sys.time() - mark

BRAAK <- rbind( BRAAK_ACC, BRAAK_PCC, BRAAK_DLPFC )
```
```{r NeuroPath_Associtation_2_1, include=FALSE, cache=TRUE}
#CERAD
Path <- 'ceradsc'
mark<-Sys.time()
CERAD_ACC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'ACC' )}, error=function(e) NULL)
Sys.time() - mark
```
```{r NeuroPath_Associtation_2_2, include=FALSE, cache=TRUE}
mark<-Sys.time()
CERAD_PCC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'PCC' )}, error=function(e) NULL)
Sys.time() - mark
```
```{r NeuroPath_Associtation_2_3, include=FALSE, cache=TRUE}

mark<-Sys.time()
CERAD_DLPFC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'DLPFC' )}, error=function(e) NULL)
Sys.time() - mark

CERAD <- rbind( CERAD_ACC, CERAD_PCC, CERAD_DLPFC )
```
```{r NeuroPath_Associtation_3_1, include=FALSE, cache=TRUE}

#COGDX
Path <- 'cogdx'
mark<-Sys.time()
cogdx_ACC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'ACC' )}, error=function(e) NULL)
Sys.time() - mark
```
```{r NeuroPath_Associtation_3_2, include=FALSE, cache=TRUE}

mark<-Sys.time()
cogdx_PCC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'PCC' )}, error=function(e) NULL)
Sys.time() - mark
```
```{r NeuroPath_Associtation_3_3, include=FALSE, cache=TRUE}

mark<-Sys.time()
cogdx_DLPFC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'DLPFC' )}, error=function(e) NULL)
Sys.time() - mark
```

```{r NeuroPath_Associtation_4_1, include=FALSE, cache=TRUE}

cogdx <- rbind( cogdx_ACC, cogdx_PCC, cogdx_DLPFC )

#DCFDX_LV
Path <- 'dcfdx_lv'
mark<-Sys.time()
dcfdx_ACC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'ACC' ) }, error=function(e) NULL)
Sys.time() - mark
```
```{r NeuroPath_Associtation_4_2, include=FALSE, cache=TRUE}

mark<-Sys.time()
dcfdx_PCC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'PCC' ) }, error=function(e) NULL)
Sys.time() - mark
```
```{r NeuroPath_Associtation_4_3, include=FALSE, cache=TRUE}
mark<-Sys.time()
dcfdx_DLPFC <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% tryCatch({NeuroPath_Calc( GN=temp[i],Path=Path, Exp=EXP, Tiss = 'DLPFC' ) }, error=function(e) NULL)
Sys.time() - mark

dcfdx <- rbind( dcfdx_ACC, dcfdx_PCC, dcfdx_DLPFC )
```

```{r NeuroPath_Finish, include=FALSE, cache=TRUE}

Cleaner <- function( DF, type ){
  colnames(DF)[2] <- 'Tissue'
  DF <- as.data.frame( DF, stringsAsFactors=F )
  DF$OR <- as.numeric(DF$OR)
  DF$Tissue <- as.factor(DF$Tissue)
  
  DF$`2.5 %` <- as.numeric(DF$`2.5 %`)
  DF$`97.5 %` <- as.numeric(DF$`97.5 %`)
  DF$PVal <- as.numeric(DF$PVal)
  DF$PVal_Adj <- NA
  for( Tis in levels(DF$Tissue) ){
    
    DF[as.character(DF$Tissue) == Tis,]$PVal_Adj <- apply( as.matrix(DF[as.character(DF$Tissue) == Tis,]$PVal), 
                                                           1,  
                                                           p.adjust, 
                                                           method='fdr', 
                                                           n=dim( DF[as.character(DF$Tissue) == Tis,])[1]
                                                          )
  }
  
  DF$NeuroPath <- type
  return(DF)
}
BRAAK <- Cleaner( BRAAK, 'BRAAK' )
CERAD <- Cleaner( CERAD, 'CERAD' )
dcfdx <- Cleaner( dcfdx, 'DCFDX' )
cogdx <- Cleaner( cogdx, 'COGDX' )
Master <- as.data.frame( rbind( BRAAK,CERAD,dcfdx,cogdx ), stringsAsFactors=F )


Master <- Master[, c( 'Gene', 'Tissue', 'OR', '2.5 %', '97.5 %', 'PVal', 'PVal_Adj', 'NeuroPath' ) ] 
colnames(Master) <- c( 'ENSG', 'Tissue', 'OddsRatio', 'CI_Lower', 'CI_Upper', 'PVal', 'PVal_Adj', 'NeuroPath_Type' ) 


print( paste0( 'Significantly associated genes by Neuropath Score: ' ))
table(Master[ Master$PVal_Adj < 0.05, ]$Tissue, Master[ Master$PVal_Adj < 0.05, ]$NeuroPath_Type) 
#print( paste0( 'Frequency of Gene Across Significance to Neuropath Score: '))
#table(table( Master[ Master$PVal_Adj < 0.05, ]$ENSG)) 
#print( paste0( 'Significantly associated genes by Neuropath Score with an Odds Ratio < 1: ',
#               table(Master[ Master$PVal_Adj < 0.05, ]$OddsRatio < 1 )['TRUE'] ))
#print( paste0( 'Significantly associated genes by Neuropath Score with an Odds Ratio > 1: ',
#               table(Master[ Master$PVal_Adj < 0.05, ]$OddsRatio > 1 )['TRUE'] ))
#print( 'Significantly associated genes by Neuropath Score with an Odds Ratio > 1 by Score Type: ')
#table(Master[ Master$PVal_Adj < 0.05, ]$OddsRatio > 1, Master[ Master$PVal_Adj < 0.05, ]$NeuroPath_Type ) 
# Store results
#write.csv(Master, file = '~/NeuroPath_Regression/Neuropathology_Scores.csv',  row.names =F)
```


```{r synapse.parameters, include=FALSE, cache=TRUE}
parentId = 'syn20798011';
activityName = 'Covariate Regression';
activityDescription = 'Covariate analysis and Regression of aligned effective counts with GRCh38 with CQN normalisation (DLPFC)';
thisFileName <- '2020_12_09_ROSMAP_RinEQGt5_TWASDiagnosis_Regress_hg38.Rmd'
# Github link
thisRepo <- githubr::getRepo(repository = "jgockley62/TWAS", ref="branch", refName='master')
thisFile <- githubr::getPermlink(repository = thisRepo, repositoryPath=paste0('code/',thisFileName))
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=TRUE}
activityName = 'Covariate Regression';
activityDescription = 'Covariate analysis and Regression of AD Diagnosis and aligned effective counts with GRCh38 with CQN normalisation (DLPFC)';
CODE <- syn_temp$store(synapseclient$Folder(name = "Full_RosMap_RinOver5", parentId = parentId))

Syns_Used <- ALL_USED_IDs
# Set annotations
all.annotations = list(
  dataType = 'mRNA',
  dataSubType = 'geneExp',
  summaryLevel = 'gene',
  assay	 = 'RNAseq',
  tissueTypeAbrv	= c('ACC','PCC','DLPFC'), 
  study = 'ROSMAP', 
  organism = 'HomoSapiens',
  consortium	= 'ROSMAP',
  normalizationStatus	= TRUE,
  normalizationType	= 'CQN',
  rnaquantification = 'RSEM',
  genomeAssemblyID = 'GRCh38'
)
# Store SVA results
write.table(RESIDUAL.SVA.GENE_EXPRESSION, file = 'ROSMAP_SVA_Diagnosis_Corrected.tsv', sep = '\t', quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='ROSMAP_SVA_Diagnosis_Corrected.tsv', name = 'Normalised, covariates, surrogate and Diagnosis variable adjusted residual expression (for eQTL)', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'residualGeneExpForeQTL'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)

# Store covariates
COVARIATES = rownameToFirstColumn(COVARIATES, 'SampleID')
write.table(COVARIATES, file = 'ROSMAP_SVA_Diagnosis_Covariates.tsv', row.names = F, sep = '\t', quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='ROSMAP_SVA_Diagnosis_Covariates.tsv', name = 'Covariates', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'covariates'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)

# Store filtered counts
PROCESSED_COUNTS$filteredExprMatrix$counts %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'ROSMAP_DLPFC_Diagnosis_Regressed_Counts.tsv', sep = '\t', row.names=F, quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='ROSMAP_DLPFC_Diagnosis_Regressed_Counts.tsv', name = 'Counts (filtered raw)', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'filteredCounts'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)

# Store Estimated counts
NEW.COUNTS %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'ROSMAP_DLPFC_Diagnosis_Regressed_eCounts.tsv', sep = '\t', row.names=F, quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='ROSMAP_DLPFC_Diagnosis_Regressed_Counts.tsv', name = 'Counts (estimated)', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'estimatedCounts'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
  
# Store logCPM
CQN.GENE_EXPRESSION$y %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'ROSMAP_DLPFC_Diagnosis_Regressed_logCPM.tsv', sep = '\t', row.names=F, quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='ROSMAP_DLPFC_Diagnosis_Regressed_logCPM.tsv', name = 'Counts (filtered logCPM)', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'filteredLCPM'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)

# Store cqn offsets
CQN.GENE_EXPRESSION$offset %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'ROSMAP_DLPFC_Diagnosis_Regressed_offset.tsv', sep = '\t', row.names=F, quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='ROSMAP_DLPFC_Diagnosis_Regressed_offset.tsv', name = 'Gene length and GC content offset', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'offset'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)

# Store design matrix
MODEL1 %>%
  rownameToFirstColumn('SampleID') %>%
  write.table(file = 'ROSMAP_DLPFC_Diagnosis_Regressed_Design.tsv', sep = '\t', row.names=F, quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='ROSMAP_DLPFC_Diagnosis_Regressed_Design.tsv', name = 'Design Matrix', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'designMatrix'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)

# Store surrogate variables
SURR.VAR %>%
  rownameToFirstColumn('SampleID') %>%
  write.table(file = 'ROSMAP_DLPFC_Diagnosis_Regressed_SurVar.tsv', sep = '\t', row.names=F, quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='ROSMAP_DLPFC_Diagnosis_Regressed_SurVar.tsv', name = 'Surrogate Variables', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'surrogateVariables'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
  
# Store residual gene expression for network analysis
RESIDUAL.NET.GENE_EXPRESSION %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'ROSMAP_DLPFC_Diagnosis_Regressed_netResidualExpression.tsv', sep = '\t', row.names=F, quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='ROSMAP_DLPFC_Diagnosis_Regressed_netResidualExpression.tsv', name = 'Normalised, covariates and Diagnosis removed residual expression (for network analysis)', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'residualGeneExpForNetAnlz'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)

stopCluster(cl)
```

```{r DelFiles, echo=T, results='hide'}
#system('rm *.tsv')
```
### R Source Code
[Github](`r thisFile`)

```{r knitmd, eval=FALSE, cache=FALSE, include=FALSE}
  reticulate::use_python("/usr/bin/python", required = TRUE)
  synapseclient <- reticulate::import("synapseclient")
  syn_temp <- synapseclient$Synapse()
  syn_temp$login()
  
  #_ Set this Variable to filter Rin > 5 _#
  RIN_Filt <- FALSE
  #_ Set this Variable to use sageseqr vals _#
  SageSeqR <- FALSE
  
  setwd("~/TWAS/code/")
  source("~/TWAS/utilityFunctions/knitfile2synapseClient.R")
  source("~/TWAS/utilityFunctions/hook_synapseMdSyntax_plot.R")

  createAndKnitToFolderEntityClient(file = "2020_12_09_ROSMAP_RinEQGt5_TWASDiagnosis_Regress_hg38.Rmd",
                                          parentId ="syn20798011",
                                          folderName = 'Full_RosMap_RinOver5')

```
