---
title: "CMC_DataCleaning"
author: 'JKG'
output: html_notebook
editor_options:
  chunk_output_type: console
---

Date of analysis update: "`r date()`"

| *Synapse ID* | *File Name* |
|  -----------------------------------------  |   ---------                      |
| syn21385083 | | 
| syn20835005 | |
| syn20835007 | |
| syn20834996 | |
| syn10887100 | |
| syn10887068 | |
| syn10887117 | |
| syn10887048 | |
| syn10887106 | |
| syn10887111 | |
| syn10887027 | |
| syn10887033 | |
| syn10887101 | |
| syn10887029 | |
| syn10887044 | |
| syn11273049 | |
| syn10887084 | |
| syn10887091 | |
| syn10887078 | |
| syn10887115 | |
| syn10887041 | |
| syn10887071 | |
| syn10887037 | |
| syn10887103 | |
| syn10887102 | |
| syn10887030 | |
| syn11273048 | |
| syn10887059 | |
| syn10887040 | |
| syn10887057 | |
| syn10887065 | |
| syn10887055 | |
| syn10887119 | |
| syn10887073 | |
| syn10887050 | |
| syn10887014 | |
| syn10887021 | |
| syn10887062 | |
| syn10887107 | |
| syn10887096 | |
| syn11273047 | |
| syn10887049 | |
| syn10887058 | |
| syn10887020 | |
| syn10887031 | |
| syn10887108 | |
| syn10887056 | |
| syn11273044 | |
| syn10887032 | |
| syn10887094 | |
| syn10887069 | |
| syn10887077 | |
| syn10887019 | |
| syn10887104 | |

```{r libs, include=FALSE}
knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  results='hide')
library(reticulate)
#BiocManager::install("Rgraphviz")
#BiocManager::install("graph")
#install.packages('CePa')
#BiocManager::install("phantasus")
```
##Before running 
Run the following in docker exec -it <CONTAINER> /bin/bash 
chmod 777 /root/
mv TWAS/ /home/<USR>/TWAS/
chown -hR <GID>:<USR> /home/<USR>/TWAS/
Create a ~/.synapseConfig file in the container, but do not push a container containing your credentials file to Docker Hub!!!

## Login to Synapse Will need to replace with your credentials
```{bash Login, echo=T, results='hide', eval=FALSE}
#source /root/.bashrc
synapse login -u <USR> -p <PSWD> --rememberMe

#Alternativly you can setup a credentials file as such:
touch ~/.synapseConfig
echo "[authentication]" >> ~/.synapseConfig
echo "username = <USR>" >> ~/.synapseConfig
echo "password = <PASWD>" >> ~/.synapseConfig
```

```{bash PLINKSetup, results='hide', eval=F, echo=TRUE}
mkdir ~/TWAS/bin
wget -q -P ~/TWAS/bin http://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20190617.zip 
unzip -d ~/TWAS/bin/ ~/TWAS/bin/plink_linux_x86_64_20190617.zip
wget -q -P ~/TWAS/bin http://s3.amazonaws.com/plink2-assets/plink2_linux_x86_64_20191015.zip
unzip -d ~/TWAS/bin/ ~/TWAS/bin/plink2_linux_x86_64_20191015.zip

rm ~/TWAS/bin/*.zip
rm ~/TWAS/bin/toy.*
rm ~/TWAS/bin/LICENSE
```

## Pull CMC Data From Synapse and Prep for Imputation - 
This data is protected, you wil be required to apply for access through PhsycEncode
```{bash Pull_WGS, results='hide', echo=TRUE, cache=TRUE}
CORES=`expr $(nproc) - 2`
#Pull Data with needed SNPS only and fiter GTEx for that SNP List
synapse get syn20835005
synapse get syn20835007
synapse get syn20834996
~/TWAS/bin/plink --threads $CORES --bfile All_CEU_ToTrainForTWAS --write-snplist

mkdir CMC_Genos
synapse get syn10537112 --recursive --downloadLocation CMC_Genos/

mkdir CMC_Filt_Genos
touch merge.txt
#Pull out Snps
for i in {1..22}
  do
    #--positions plink.snplist 
    bcftools norm --threads $CORES --rm-dup snps -Oz CMC_Genos/CMC_chr$i.dose.vcf.gz > CMC_Filt_Genos/Filt_CMC_chr$i.dose.vcf.gz
    vcftools --gzvcf CMC_Filt_Genos/Filt_CMC_chr$i.dose.vcf.gz --plink --out CMC_Filt_Genos/Init_Filtered_chr$i
    echo 'CMC_Filt_Genos/Init_Filtered_chr'$i'.ped CMC_Filt_Genos/Init_Filtered_chr'$i'.map' >> merge.txt
  done
grep -v 'chr1.ped' merge.txt > foo
mv foo merge.txt

#Merge Genotype
~/TWAS/bin/plink --threads $CORES --file CMC_Filt_Genos/Init_Filtered_chr1 --merge-list merge.txt --make-bed --out Merged_Initial_CMC_Geno

#Remove old Map/Ped Files
rm CMC_Filt_Genos/Init_Filtered_chr*.map
rm CMC_Filt_Genos/Init_Filtered_chr*.ped
```

### ReName the MSBB SNPS so the can be combined
```{python ReNamer_ONE, echo=F, results='hide',python.reticulate=T}
import sys
import os
import re

InFile = 'All_CEU_ToTrainForTWAS.bim'
InFileb = 'Merged_Initial_CMC_Geno.bim' #LDREF_FiltSNPs.bim

RS_IDs = {}
RS_Pos = {}

F1 = open(InFile)
for line in F1:
	LINE = line.rstrip('\r\n')
	lst = LINE.split('\t')
	RS_IDs[ ''.join([lst[0], '_', lst[3]]) ] = lst[1]
	RS_Pos[ ''.join([lst[0], '_', lst[3]]) ] = lst[2]
F1.close()

#Open Output
OUT = open( ''.join([ InFileb.replace(".bim", ""), 'Renamed.bim']), 'w' )

F1 = open(InFileb)
for line in F1:
	LINE = line.rstrip('\r\n')
	lst = LINE.split('\t')
	if( ''.join([ lst[0], "_", lst[3] ]) in RS_IDs):
		Name = RS_IDs[ ''.join([ lst[0], "_", lst[3] ])]
		Entry = '\t'.join([ lst[0], Name, lst[2], lst[3], lst[4], lst[5] ])
		print >>OUT, Entry
	else:
		print >>OUT, LINE

F1.close()
OUT.close()
```

## Pull Imputation SNPS - 
```{bash Pull_impute, results='hide', echo=TRUE, cache=TRUE}
CORES=`expr $(nproc) - 2`

mv Merged_Initial_CMC_GenoRenamed.bim Merged_Initial_CMC_Geno.bim
~/TWAS/bin/plink --threads $CORES --bfile Merged_Initial_CMC_Geno --extract plink.snplist --make-bed --out CMC_Genos_For_Impute

```

```{r synapse.parameters.head, include=FALSE, results='hide', echo=FALSE, cache=TRUE}
parentId = 'syn18936948';
activityName = 'Processed CMC Genotypes TWAS';
activityDescription = 'Proccessed and Filtered CMC Genotype Data';
thisFileName <- 'CMC_DataProcessing.Rmd'
# Github link
thisRepo <- githubr::getRepo(repository = "jgockley62/TWAS", ref="branch", refName='master')
thisFile <- githubr::getPermlink(repository = thisRepo, repositoryPath=paste0('code/',thisFileName))
```

```{r synapse.parameters, include=FALSE, results='hide', echo=FALSE, cache=TRUE}

activityName = 'Processed CMC Genotypes TWAS';
activityDescription = 'Proccessed and Filtered CMC Genotype Data';

CODE <- syn_temp$store(synapseclient$Folder(name = "CMC_FilteredGenotypes", parentId = parentId))
#CODEalt <- syn_temp$store(synapseclient$Folder(name = "CMC_FilteredGenotypes", parentId = parentId))

#Set Used SynIDs For Provenance
Syns_Used <- c("syn21385083", "syn20835005", "syn20835007", "syn20834996", "syn10887100", "syn10887068", "syn10887117", "syn10887048", "syn10887106", "syn10887111", "syn10887027", "syn10887033", "syn10887101", "syn10887029", "syn10887044", "syn11273049", "syn10887084", "syn10887091", "syn10887078", "syn10887115", "syn10887041", "syn10887071", "syn10887037", "syn10887103", "syn10887102", "syn10887030", "syn11273048", "syn10887059", "syn10887040", "syn10887057", "syn10887065", "syn10887055", "syn10887119", "syn10887073", "syn10887050", "syn10887014", "syn10887021", "syn10887062", "syn10887107", "syn10887096", "syn11273047", "syn10887049", "syn10887058", "syn10887020", "syn10887031", "syn10887108", "syn10887056", "syn11273044", "syn10887032", "syn10887094", "syn10887069", "syn10887077", "syn10887019", "syn10887104"
)
 
# Set annotations
Var.all.annotations = list(
  dataType = 'Variant',
  dataSubType = 'SNP',
  summaryLevel = 'Nucleotide',
  assay  = 'Genotype Array',
  tissueTypeAbrv  = c('Mixed Tissues'), 
  study = 'CMC', 
  organism = 'HomoSapiens',
  consortium  = 'CMC',
  genomeAssemblyID = 'GRCh37'
)

# Set annotations
EXP.all.annotations = list(
  dataType = 'mRNA',
  dataSubType = 'geneExp',
  summaryLevel = 'gene',
  assay  = 'RNAseq',
  tissueTypeAbrv  = c('Multiple Brain'), 
  study = 'CMC', 
  organism = 'HomoSapiens',
  consortium  = 'CMC',
  normalizationStatus = TRUE,
  normalizationType = 'Scaled TPM',
  rnaquantification = 'UNK',
  genomeAssemblyID = 'GRCh38'
)

Basic.annotations = list(
  dataType = 'MetaData',
  dataSubType = 'CMC Sample Meta Data',
  study = 'CMC', 
  organism = 'HomoSapiens',
  consortium  = 'CMC'
)

ENRICH_OBJ <- syn_temp$store( synapseclient$File( path='CMC_Genos_For_Impute.bim', name = 'CMC Genotypes For Imputing exp BIM', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
syn_temp$setAnnotations(ENRICH_OBJ, annotations = Var.all.annotations)

ENRICH_OBJ <- syn_temp$store( synapseclient$File( path='CMC_Genos_For_Impute.fam', name = 'CMC Genotypes For Imputing exp FAM', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
syn_temp$setAnnotations(ENRICH_OBJ, annotations = Var.all.annotations)

ENRICH_OBJ <- syn_temp$store( synapseclient$File( path='CMC_Genos_For_Impute.bed', name = 'CMC Genotypes For Imputing exp BED ', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
syn_temp$setAnnotations(ENRICH_OBJ, annotations = Var.all.annotations)
```

## Clean

```{bash cleanup, results='hide', echo=TRUE, cache=TRUE}
rm *.fam
rm *.bed
rm *.bim
rm *.fam
rm *.nosex
rm *.ped
rm *.map
rm *.missnp
rm *.log
rm *.snplist
rm merge.txt
rm -r CMC_Genos/
rm -r CMC_Filt_Genos/
```

### R Source Code
[Github](`r thisFile`)

```{r knitmd, eval=FALSE, cache=FALSE, include=FALSE}
synapseclient <- reticulate::import("synapseclient")
syn_temp <- synapseclient$Synapse()
syn_temp$login()

setwd("~/TWAS/code/")
source("~/TWAS/utilityFunctions/knitfile2synapseClient.R")
source("~/TWAS/utilityFunctions/hook_synapseMdSyntax_plot.R")

createAndKnitToFolderEntityClient(file = "CMC_DataProcessing.Rmd",
                                          parentId ="syn18936948",
                                          folderName = 'CMC_FilteredGenotypes')

```
