---
title: "Plot Results"
output: html_document
---
---
title: "Manhattan and qqplot results"
author: "JKG"
date: "07/02/2020"
output: pdf_document
  chunk_output_type: console
---

```{r libs, include=FALSE}
library(qqman)
library(ggplot2)
library(ggrepel)
library(dplyr)

knitr::opts_chunk$set(
  #python.reticulate=FALSE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  results='hide')

```

```{r Prep, include=FALSE}

Values <- read.table( syn_temp$get('syn21529460')$path, header = T, sep='\t' )
Assoc <- read.table( syn_temp$get('syn21577463')$path, header = T, sep='\t' )
row.names(Values) <- Values$probeID

Assoc$Gene <- Assoc$ID
GenesOfInterest <- c("MTCH2",
                     "EED",
                     "KNOP1",
                     "CD2AP",
                     "APOC1",
                     "CLPTM1",
                     "TREM2",
                     "CEACAM19")
JCP <- c("PVRL2",
         "PVR",
         "TMEM223",
         "NUP160",
         "PTPRJ",
         "C1QTNF4")

JCPval <- c( "POLR2E",
             "POLR2G",
             "ZNF660",
             "MADD"
)

Assoc<-Assoc[,(colnames(Assoc) %in% 'PANEL')==F]



Assoc<-Assoc[complete.cases(Assoc),]
Assoc$TWAS.P <- as.numeric(as.character(Assoc$TWAS.P))
Assoc$FDR_TWAS.P <- p.adjust( Assoc$TWAS.P, method='fdr', dim(Assoc)[1])
don <- Assoc %>% 
  
  # Compute chromosome size
  group_by(CHR) %>% 
  summarise(chr_len=max(P0)) %>% 
  
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(as.numeric(chr_len))-as.numeric(chr_len)) %>%
  select(-chr_len) %>%
  
  # Add this info to the initial dataset
  left_join(Assoc, ., by=c("CHR"="CHR")) %>%
  
  # Add a cumulative position of each SNP
  arrange(CHR, P0) %>%
  mutate( BPcum=P0+tot) %>%
  
  # Add highlight and annotation information
  mutate( is_highlight=ifelse(Gene %in% GenesOfInterest, "yes", 
                                ifelse(Gene %in% JCP, "drop", 
                                       ifelse(Gene %in% JCPval, "NS", "no")) )) %>%
  
  mutate( is_annotate=ifelse( FDR_TWAS.P < 0.05 , "yes", "no")) 

# Prepare X axis
axisdf <- don %>% group_by(CHR) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )

#don$is_annotate <- don$is_highlight
```

```{r Plot}
# Make the plot
ggplot(don, aes(x=BPcum, y=-log10(TWAS.P))) +
  
  #AddSigLine
  geom_hline(yintercept = -log10( max(Assoc[Assoc$FDR_TWAS.P < 0.05,]$TWAS.P) )) +
  #p.adjust(7.467145e-06, method='fdr', dim(Assoc)[1] )
  
  # Show all points
  geom_point( aes(color=as.factor(CHR)), alpha=0.8, size=1.3) +
  scale_color_manual(values = rep(c("grey", "black"), 22 )) +
  
  # custom X axis:
  scale_x_continuous( name="Chromosome", label = axisdf$CHR, breaks= axisdf$center ) +
  scale_y_continuous(name="-Log10( TWAS Association P-Value )", expand = c(0, 0), limits=c(0,38) ) +     # remove space between plot area and x axis
  
  # Add highlighted points
  geom_point(data = subset(don, is_highlight=="yes"), color="orange", size=2) +
  geom_point(data = subset(don, is_highlight=="drop"), color="lightslateblue", size=2) +
  geom_point(data = subset(don, is_highlight=="NS"), color="mediumturquoise", size=2) +
  
  # Add label using ggrepel to avoid overlapping
  geom_label_repel( data=subset(don, is_annotate=="yes"), aes(label=Gene), size=2) +
  
  # Custom the theme:
  theme_bw() +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  )

qq(Assoc$TWAS.P, bty='n', las=1)
```

```{r SavePlot,include=FALSE}
# Make the plot
pdf(file = "TWAS_Manhattan.pdf")

  ggplot(don, aes(x=BPcum, y=-log10(TWAS.P))) +
    
    #AddSigLine
    geom_hline(yintercept = -log10( max(Assoc[Assoc$FDR_TWAS.P < 0.05,]$TWAS.P) )) +
    #p.adjust(7.467145e-06, method='fdr', dim(Assoc)[1] )
    
    # Show all points
    geom_point( aes(color=as.factor(CHR)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "black"), 22 )) +
    
    # custom X axis:
    scale_x_continuous( name="Chromosome", label = axisdf$CHR, breaks= axisdf$center ) +
    scale_y_continuous(name="-Log10( TWAS Association P-Value )", expand = c(0, 0), limits=c(0,38) ) +     # remove space between plot area and x axis
    
    # Add highlighted points
    geom_point(data = subset(don, is_highlight=="yes"), color="orange", size=2) +
    geom_point(data = subset(don, is_highlight=="drop"), color="lightslateblue", size=2) +
    geom_point(data = subset(don, is_highlight=="NS"), color="mediumturquoise", size=2) +
    
    # Add label using ggrepel to avoid overlapping
    geom_label_repel( data=subset(don, is_annotate=="yes"), aes(label=Gene), size=2) +
    
    # Custom the theme:
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )
dev.off()

p <- ggplot(don, aes(x=BPcum, y=-log10(TWAS.P))) +
    
    #AddSigLine
    geom_hline(yintercept = -log10( max(Assoc[Assoc$FDR_TWAS.P < 0.05,]$TWAS.P) )) +
    #p.adjust(7.467145e-06, method='fdr', dim(Assoc)[1] )
    
    # Show all points
    geom_point( aes(color=as.factor(CHR)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "black"), 22 )) +
    
    # custom X axis:
    scale_x_continuous( name="Chromosome", label = axisdf$CHR, breaks= axisdf$center ) +
    scale_y_continuous(name="-Log10( TWAS Association P-Value )", expand = c(0, 0), limits=c(0,38) ) +     # remove space between plot area and x axis
    
    # Add highlighted points
    geom_point(data = subset(don, is_highlight=="yes"), color="orange", size=2) +
    geom_point(data = subset(don, is_highlight=="drop"), color="lightslateblue", size=2) +
    geom_point(data = subset(don, is_highlight=="NS"), color="mediumturquoise", size=2) +
    
    # Add label using ggrepel to avoid overlapping
    geom_label_repel( data=subset(don, is_annotate=="yes"), aes(label=Gene), size=2) +
    
    # Custom the theme:
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )
ggsave("TWAS_Manhattan.eps",p)

pdf(file = "TWAS_qq.pdf")
qq(Assoc$TWAS.P, bty='n', las=1)
dev.off()

setEPS()
postscript("TWAS_qq.eps", onefile = TRUE)
qq(Assoc$TWAS.P, bty='n', las=1)
dev.off()

```


```{r synapse.parameters, include=FALSE, cache=TRUE}
parentId = 'syn18936948';
activityName = 'TWAS Results';
activityDescription = 'Results Plots from TWAS Analysis';
thisFileName <- 'ResultsPlots.Rmd'
# Github link
thisRepo <- githubr::getRepo(repository = "jgockley62/TWAS", ref="branch", refName='master')
thisFile <- githubr::getPermlink(repository = thisRepo, repositoryPath=paste0('code/',thisFileName))
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
activityName = 'TWAS Results';
activityDescription = 'Results Plots from TWAS Analysis';
CODE <- syn_temp$store(synapseclient$Folder(name = "TWAS Results Figures", parentId = parentId))

#Set Used SynIDs For Provenance
#Syns_Used <- c("")

# Set annotations
all.annotations = list(
  dataType = 'pdf',
  dataSubType = 'geneExp',
  summaryLevel = 'gene',
  assay	 = 'TWAS',
  study = 'TWAS', 
  organism = 'HomoSapiens',
  consortium	= c( 'ROSMAP', 'MSBB','Mayo' ),
  normalizationStatus	= TRUE,
  normalizationType	= 'CQN',
  genomeAssemblyID = 'GRCh37'
)

#TODO add provenance used = Syns_Used,

# PDFs
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='TWAS_Manhattan.pdf', name = 'Twas Gene Association Manhattan Plot', parentId=CODE$properties$id ), activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'Manhattan Plot'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
  
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='TWAS_qq.pdf', name = 'Twas Gene Association QQ-Plot', parentId=CODE$properties$id ), activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'QQ-Plot'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
  
# PDFs
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='TWAS_Manhattan.eps', name = 'Twas Gene Association Manhattan Plot EPS', parentId=CODE$properties$id ), activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataType = "eps"
  all.annotations$dataSubType = 'Manhattan Plot'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
  
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='TWAS_qq.eps', name = 'Twas Gene Association QQ-Plot EPS', parentId=CODE$properties$id ), activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataType = "eps"
  all.annotations$dataSubType = 'QQ-Plot'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
  
```

```{r knitmd, eval=FALSE, cache=FALSE, include=FALSE}
synapseclient <- reticulate::import("synapseclient")
syn_temp  <- synapseclient$Synapse()
syn_temp$login()

setwd("~/TWAS/code/")
source("~/TWAS/utilityFunctions/knitfile2synapseClient.R")
source("~/TWAS/utilityFunctions/hook_synapseMdSyntax_plot.R")

createAndKnitToFolderEntityClient(file = "ResultsPlots.Rmd",
                                          parentId ="syn18936948",
                                          folderName = 'TWAS Results Figures')
```
