---
  title: "Summary-data-based Mendelian Randomization"
author: "Jake Gockley"
date: "01/17/2019"
header-includes:
  - \usepackage{multicol}
  - \newcommand{\btwocol}{\begin{multicols}{2}}
  - \newcommand{\etwocol}{\end{multicols}}
output:
  html_document: default
pdf_document: default
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
library(readr)
library(snpStats)
library(sem)
library(parallel)
```

```{r code, eval=TRUE, cache=FALSE, include=TRUE}

#Pull eQTL Data
system( 'synapse get syn16984815' )

#Pull Weights
#system('aws s3 cp s3://jkg-s3-synapseencryptedexternalbucket-zszdd03ghnb2/ROSMAP/TWAS_2019_12_02_PanCortex_DiagnosisRegressed_Weights/ All_Data/ALL/ --recursive')
#system('rm All_Data/ALL/*.hsq')
system('synapse get -r syn21363795 --downloadLocation All_Data/ALL/ ')
system('rm All_Data/ALL/*.hsq')

#PullGenos
system('synapse get syn20820120 --downloadLocation Binary_Cleaned/')
system('synapse get syn20820121 --downloadLocation Binary_Cleaned/')
system('synapse get syn20820122 --downloadLocation Binary_Cleaned/')

#GWAS
system('synapse get syn18999109 --downloadLocation .')

#Expression
system('synapse get syn21011186 --downloadLocation .')
Exp <- read.table('ALL_CEU_RNA_Seq_SVAadusted_DiagnosisRegressed.tsv', header = T, sep='\t')

#LoadKunkle
Kunk <- read.table('Kunkle_etal_2019_IGAP_Summary_statistics.with_allelefreqs.txt', header =T, sep="")

#Try the MR for one gene
load("All_Data/ALL/ENSG00000130208.wgt.RDat")
PLINK <- read.plink(bed = 'Binary_Cleaned/CEU_AncestryMatched_Genotypes.bed' ,
                    bim = 'Binary_Cleaned/CEU_AncestryMatched_Genotypes.bim',
                    fam = 'Binary_Cleaned/CEU_AncestryMatched_Genotypes.fam' 
)

#Filter Kunkle for relevant SNPS by their IDS
K <- Kunk[ Kunk$MarkerName %in% PLINK$map$snp.name, ]
row.names(K) <- K$MarkerName
mygwas<-K[,c(3:5,9,6:8)]
colnames(mygwas) <- c("SNP", "A1", 'A2', "freq", "b", "se", "p" )
mygwas$n <- rep( 2003, dim(mygwas)[1] )

write.table(snps$V2, 'snp.list', row.names=F, col.names=F, quote=F, sep='\t')
write.table(mygwas, 'mygwas.ma', row.names=F, col.names=T, quote=F, sep='\t')


if('plink' %in% list.files() ){}else{
system('wget http://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20200107.zip')
system('unzip plink_linux_x86_64_20200107.zip')
}
if('smr_Linux' %in% list.files() ){}else{
system('wget https://cnsgenomics.com/software/smr/download/smr_Linux.zip')
system('unzip smr_Linux.zip')
}
system( './plink -bfile Binary_Cleaned/CEU_AncestryMatched_Genotypes --extract snp.list --make-bed --out Temp_plink')
system( './smr_Linux --bfile Temp_plink --gwas-summary mygwas.ma --beqtl-summary myeqtl --out mymulti --smr-multi')

#make the eqtl data into BESD format
EQTL <- read.csv( 'Cortex_MetaAnalysis_ROSMAP_CMC_HBCC_Mayo_cis_eQTL_release.csv', header = T )

EQTl <- EQTL[ as.character(EQTL$snpid ) %in% as.character(PLINK$map$snp.name) , ]
EQtl <- EQTl[ as.character(EQTl$gene ) %in% as.character(Exp$TargetID) , ]

EQTL_sink <- EQTL
EQTL <- EQtl

#SNP gene beta t-stat p-value FDR
EQTL$geneSymbol <- as.character(EQTL$geneSymbol)
EQTL$gene <- as.character(EQTL$gene)
EQTL[ as.character(EQTL$geneSymbol) =="",]$geneSymbol <- EQTL[ as.character(EQTL$geneSymbol) =="",]$gene

EQTL$snpid <- as.character(EQTL$snpid)
EQTL$DUP <- paste0( as.character(EQTL$snpid), '_', as.character(EQTL$gene))
EQTL_deDup <- EQTL[ !duplicated(EQTL$DUP), ]

##LOAD EXPRESSION AND BIM
###FILTER FOR RELEVANT SNP-GENEs
##### wExp <- read.table(file = 'WeightList/MaterWeight.lst', sep = "", header = T)
#EQTL_deDup <- EQTL_deDup[ EQTL_deDup$snpid %in% as.character(PLINK$bim[,2]), ]
#EQTL_deDup <- EQTL_deDup[ EQTL_deDup$gene %in% as.character(Exp[,2]), ]

Repo <- function( n ){
  #'@Mat eg. EQTL_deDup 
  #'@n eg a number
  if( EQTL_deDup[n,]$geneSymbol == ""){
    EQTL_deDup[n,]$geneSymbol <- EQTL_deDup[n,]$gene
  }else{}
}

REPS <- row.names( EQTL_deDup[ EQTL_deDup$geneSymbol == "", ] )

eqtl <- EQTL_deDup[ , c('snpid', 'geneSymbol', 'beta', 'statistic', 'pvalue', 'beta') ]
colnames(eqtl) <- c( 'SNP', 'gene', 'beta', 't-stat', 'p-value', 'FDR' )
write.table( eqtl, file = 'mateQTL_geneN.txt', col.names = T, row.names=F, quote=F, sep='\t' )

Eeqtl <- EQTL_deDup[ , c('snpid', 'gene', 'beta', 'statistic', 'pvalue', 'beta') ]
colnames(Eeqtl) <- c( 'SNP', 'gene', 'beta', 't-stat', 'p-value', 'FDR' )
write.table( Eeqtl, file = 'mateQTL_ensgN.txt', col.names = T, row.names=F, quote=F, sep='\t' )

system( './smr_Linux --eqtl-summary mateQTL_ensgN.txt --matrix-eqtl-format --make-besd --out mybesd_ensgN' )
system( './smr_Linux --eqtl-summary mateQTL_geneN.txt --matrix-eqtl-format --make-besd --out mybesd_geneN' )

##Fix the ,esi and .epi files
AleleSelect <- function( Row ){
  if( Row[4] > 0 ){
    Eff <- as.character(Row[9])
    Unaff <- as.character(c(Row[6],Row[7])[ (c(Row[6],Row[7]) %in% Eff) == F ])
  }else{
    Unaff <- as.character(Row[9])
    Eff <- as.character(c(Row[6],Row[7])[ (c(Row[6],Row[7]) %in% Unaff) == F ])
  }
  if(Eff == Row[7]){
    Freq <- Row[8]
  }else{
    Freq <- 1-as.numeric(Row[8])
  }
  
  return( c(as.character(Eff),as.character(Unaff),as.character(Freq)) )
}

ESiProcess <- function( Fil ){
  #'@Fil the loaded .esi file (ie File)
  ESi_info <- EQTL_deDup[ !duplicated(EQTL_deDup$snpid), c("chromosome", "snpLocation", "snpid", "statistic", "beta", "A1", "A2", "A2freq", "expressionIncreasingAllele")]
  row.names(ESi_info) <- ESi_info$snpid
  Fil$V1 <- as.character(ESi_info[ as.character( Fil$V2 ), ]$chromosome)
  Fil$V4 <- as.character(ESi_info[ as.character( Fil$V2 ), ]$snpLocation)
  
  ESi_info$A1 <- as.character(ESi_info$A1)
  ESi_info$A2 <- as.character(ESi_info$A2)
  ESi_info$expressionIncreasingAllele <- as.character(ESi_info$expressionIncreasingAllele)
  
  temp <- t( apply(ESi_info, 1, AleleSelect) )
  Fil[,5:7] <- temp[ as.character(Fil$V2),]
  return(Fil)
}

File <- read.table(file='mybesd_ensgN.esi', header=F, sep='\t')
File2 <- ESiProcess( File )
write.table(File2, file='mybesd_ensgN.esi', col.names=F, row.names = F, quote=F, sep='\t')

File <- read.table(file='mybesd_geneN.esi', header=F, sep='\t')
File2 <- ESiProcess( File )
write.table(File2, file='mybesd_geneN.esi', col.names=F, row.names = F, quote=F, sep='\t')

#ESPI
row.names(Exp) <- as.character(Exp$TargetID)
Exp$TargetID <- as.character(Exp$TargetID)
Exp$Gene_Symbol <- as.character(Exp$Gene_Symbol)

EQTL_deDup$gene <- as.character(EQTL_deDup$gene)
EQTL_deDup$geneSymbol <- as.character(EQTL_deDup$geneSymbol)
EQTL_deDup$chromosome <- as.character(EQTL_deDup$chromosome)
EQTL_deDup$strand <- as.character(EQTL_deDup$strand)

#Split EQTL_deDup into special data frames to reference
EQTL_deDup_GN <- EQTL_deDup[ !duplicated(EQTL_deDup$geneSymbol), ]
EQTL_deDup_ENSG <- EQTL_deDup[ !duplicated(EQTL_deDup$gene), ]
  
ESpProcess <- function( Name ){
  #'@Name row with gene name or ENSG in column 2
  if(Name[2] %in% EQTL_deDup_ENSG$gene){
    Row <- EQTL_deDup_ENSG[ EQTL_deDup_ENSG$gene %in% Name[2], ][1,]
  }else{
    Row <- EQTL_deDup_GN[ EQTL_deDup_GN$geneSymbol %in% Name[2], ][1,]
  }
  st <- c('+', '-')
  names(st) <- c("1","-1")
  
  return( c( as.character(Row[1]), 
             as.character(Name[2]), 
             as.character("0"), 
             as.character(Exp[ as.character(Row[5]), 4]),
             as.character(Name[2]),
             as.character(st[Row[,15]])  
             ) 
          )
}

File <- read.table(file='mybesd_geneN.epi', header=F, sep='\t')
File$V2 <- as.character(File$V2)
foo <- t( apply(File, 1, ESpProcess ) )
write.table( foo, file='mybesd_geneN.epi', row.names=F, col.names=F, quote = F, sep='\t' )

File <- read.table(file='mybesd_ensgN.epi', header=F, sep='\t')
File$V2 <- as.character(File$V2)
foo <- t( apply(File, 1, ESpProcess ) )
write.table( foo, file='mybesd_ensgN.epi', row.names=F, col.names=F, quote = F, sep='\t' )


Genes <- c("APOC1", "EED", "CD2AP", "CEACAM19", "CLPTM1", "MTCH2", "TREM2", "KNOP1", "POLR2E", "POLR2G", "ZNF660", "MADD", "PVRL2", "PVR", "TMEM223", "NUP160", "PTPRJ", "C1QTNF4")
ENSG <- c

write.table(EQTL_deDup_GN[ EQTL_deDup_GN$geneSymbol %in% Genes, ]$gene, "myprobe.list", row.names=F, col.names=F, quote=F, sep='\t')


library(doParallel)
library(foreach)

#Start Master File
system( 'echo "probeID ProbeChr  Gene  Probe_bp  topSNP  topSNP_chr  topSNP_bp A1  A2  Freq  b_GWAS  se_GWAS tp_GWAS b_eQTL  se_eQTL p_eQTL  b_SMR se_SMR  p_SMR p_SMR_multi p_HEIDI nsnp_HEIDI" > Compiled_SMR.tsv')

#List <- read.table( 'WeightList/MaterWeight.lst', header = T, sep='')
List <- cbind(list.files('All_Data/ALL/'), paste0( 'All_Data/ALL/', list.files('All_Data/ALL/') ))
List[,1] <- gsub('.wgt.RDat','',List[,1])
List <- as.data.frame(List)
colnames(List) <- c('ID','WGT')
#i <- 1
#List <- as.data.frame(List)
#colnames(List) <- 'WGT'
Runner <- function( i ){
  #Load Gene Info
  load(as.character( List$WGT )[i])
  Probe <- as.character( List$ID )[i]
  SNPs <- snps$V2
  #Save Temp Files
  write.table(Probe, paste0("myprobe_", Probe, ".list"), row.names=F, col.names=F, sep="\t", quote=F)
  write.table(SNPs, paste0("mysnp_", Probe, ".list"), row.names=F, col.names=F, sep="\t", quote=F)
  #Run Model
  system( paste0("./smr_Linux --bfile Binary_Cleaned/CEU_AncestryMatched_Genotypes --gwas-summary mygwas.ma --thread-num 1 --beqtl-summary mybesd_ensgN --ld-upper-limit 1 --out temp_Ensg_", Probe, " --extract-snp mysnp_", Probe, ".list --extract-probe myprobe_", Probe, ".list --peqtl-heidi 0 --ld-multi-snp 1 --smr-multi --peqtl-smr 1 --set-wind 2000 "))
  #Pull Info
  system( paste0("grep -v 'probeID' temp_Ensg_", Probe, ".msmr >> Compiled_SMR.tsv") )
  #remove Temp Info
  system( paste0("rm myprobe_", Probe, ".list") )
  system( paste0("rm mysnp_", Probe, ".list") )
  system( paste0("rm temp_Ensg_", Probe, ".*") )
}

cl <- makeCluster(  parallel::detectCores()-2 )
registerDoParallel(cl)
foo <- Sys.time()
  foreach(i=1:length(List$ID)[1]) %dopar% Runner( i )
Sys.time()-foo
stopCluster(cl)

Table<-read.table(file='Compiled_SMR.tsv', sep='\t', header=T)

system( 'awk -F \',\' \'{ print $5"\t"$6 }\' Cortex_MetaAnalysis_ROSMAP_CMC_HBCC_Mayo_cis_eQTL_release.csv | sort | uniq | grep -v \'geneSymbol\' > Translate.txt')
Trans <- read.table( file='Translate.txt', sep ="\t", header=F)
row.names(Trans) <- Trans$V1
Trans$V1 <- as.character(Trans$V1) 
Trans$V2 <- as.character(Trans$V2) 

Trans[ is.na(Trans$V2), ]$V2 <- Trans[ is.na(Trans$V2), ]$V1
Trans[ Trans$V2 == "", ]$V2 <- Trans[ Trans$V2 == "", ]$V1

Table$Gene <- as.character(Table$Gene)
Table$probeID <- as.character(Table$probeID)

Table$Gene <- Trans[Table$probeID,]$V2
write.table(Table, file='Final_SMR_TWAS_Validation.tsv', row.names=F, col.names=T, quote=F, sep ='\t')
```


```{r synapse.parameters, include=FALSE, cache=TRUE}
parentId = 'syn18936948';
activityName = 'Mendelian Randomization';
activityDescription = 'Validation of TWAS associations via SMR analysis';
thisFileName <- 'SMR.Rmd'
# Github link
thisRepo <- githubr::getRepo(repository = "jgockley62/TWAS", ref="branch", refName='master')
thisFile <- githubr::getPermlink(repository = thisRepo, repositoryPath=paste0('code/',thisFileName))
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
activityName = 'Mendelian Randomization';
activityDescription = 'Validation of TWAS associations via SMR analysis';
CODE <- syn_temp$store(synapseclient$Folder(name = "Summary-data-based Mendelian Randomization", parentId = parentId))

# Set annotations
all.annotations = list(
  dataType = 'analysis',
  dataSubType = 'SMR_Validation_Tests',
  summaryLevel = 'gene',
  assay	 = 'TWAS',
  tissueTypeAbrv	= c('NA'), 
  study = c('ROSMAP','MSBB', 'Mayo'),
  organism = 'HomoSapiens',
  consortium	=  c('ROSMAP','MSBB', 'Mayo') ,
  normalizationStatus	= TRUE,
  normalizationType	= 'NA',
  rnaquantification = 'NA',
  genomeAssemblyID = 'hg19'
)

# Store SMR
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='Final_SMR_TWAS_Validation.tsv', name = 'SMR Values', parentId=CODE$properties$id ), activityName = activityName, executed = thisFile, activityDescription = activityDescription)
all.annotations$dataSubType = 'Summary Stats'
syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
```

### R Source Code
[Github](`r thisFile`)

```{r knitmd, eval=FALSE, cache=FALSE, include=FALSE}
reticulate::use_python("/usr/bin/python", required = TRUE)
synapseclient <- reticulate::import("synapseclient")
syn_temp <- synapseclient$Synapse()
syn_temp$login()

setwd("~/TWAS/code/")
source("~/TWAS/utilityFunctions/knitfile2synapseClient.R")
source("~/TWAS/utilityFunctions/hook_synapseMdSyntax_plot.R")
createAndKnitToFolderEntityClient(file = "SMR.Rmd",
                                  parentId ="syn18936948",
                                  folderName = 'Summary-data-based Mendelian Randomization')
```